---
title: "Anaquin - Vignette"
author: "Ted Wong (t.wong@garvan.org.au)"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: pdf_document
vignette: >
  %\VignetteIndexEntry{Anaquin - Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
type: article-journal
linkcolor: blue
urlcolor: blue
citecolor: blue
---

# Citation

[1] S.A Hardwick. Spliced synthetic genes as internal controls in RNA sequencing experiments. *Nature Methods*, 2016.

# Need help?

Visit our website to learn more about sequins in quantifying your experiment: [www.sequin.xyz](www.sequin.xyz).

We have an active Slack channel, [sequins.slack.com](sequins.slack.com), where you can chat directly with our team. Please email us at anaquin@garvan.org.au for an invitation. You can also post your questions on [www.biostars.org](www.biostars.org) and [www.seqanswers.com](www.seqanswers.com), we will answer your questions as we actively monitor the sites. 

# Overview

In this document, we show how to conduct statistical analysis that models the performance of synthetic sequin controls in next-generation-sequencing (NGS) experiment. The controls can be used for RNA-Seq analysis, e.g., normalization, gene expression, differential analysis, and other applications. We call the sequins **RnaQuin** for *RNA-Seq sequins*, and the statistical framework **Anaquin**.

This vignette is written for R-usage. However, Anaquin is a framework covering the entire RNA-Seq workflow; reads alignment, transcriptome assembly, gene expression and differential analysis, etc. Consequently, the R-package (and it's documentation) is a subset of the overall Anaquin framework. We also distribute a workflow guide with this vigentte on how Anaquin can be used in a bioinformatics pipeleine. The documentation cane be found on Bioconductor and on our website.

It is important to note Anaquin is both command-line tool and R-package. Our detailed guide has details on how the command-line tool can be used with the R-package.

# Sequins

Next-generation sequencing (NGS) enables rapid, cheap and high-throughput determination of sequences within a userâ€™s sample. NGS methods have been applied widely, and have fuelled major advances in the life sciences and clinical health care over the past decade. However, NGS typically generates a large amount of sequencing data that must be first analyzed and interpreted with bioinformatic tools. There is no standard way to perform an analysis of NGS data; different tools provide different advantages in different situations. The complexity and variation of sequences further compound this problem, and there is little reference by which compare next-generation sequencing and analysis.

To address this problem, we have developed a suite of synthetic nucleic-acid sequins (sequencing spike-ins). Sequins are fractionally added to the extracted nucleic-acid sample prior to library preparation, so they are sequenced along with your sample of interest. We can use the sequins as an internal quantitative and qualitative control to assess any stage of the next-generation sequencing workflow.

Why do we need sequins?

  - Quantification of accuracy; linear model between input concentration and measured variables
  - Derive normalization factor; sequins serve as negative controls
  - Construct receiver operating characteristic (ROC) that tabulates sensitivity
  - Determine empharical limit of quantification; below which measurement becomes increasingly stochastic  

# Mixture

Sequins are combined together across a range of concentrations to formulate a mixture. Mixture file (CSV) is a text file that specifies the concentration of each sequin within a mixture. Mixture files are often required as input to enable Anaquin to perform quantitative analysis. Mixture file can be downloaded from our website:

[www.sequin.xyz](www.sequin.xyz )

Let's demostrate RnaQuin mixture A with a simple example. Load the mixture file (you can also download the file directly from our website):

```{r}
library('Anaquin')
data(mixtureA)
head(mixtureA)
```

Each row represents a sequin. *Column ID* gives the sequin names, *Length* is the length of the sequin in nucleotide bases, 'MXA' gives the concentration level in attom/ul.

Imagine we have two RNA-Seq experiments; a well-designed experiment and a poorly-designed experiment We would like to quantify their expression accuracy.

Let's simulate the experiments:

```{r}
set.seed(1234)
sim1 <- 1.0 + 1.2*log2(mixtureA$MXA) + rnorm(nrow(mixtureA),0,1)
sim2 <- c(1.0 + rnorm(100,1,3), 1.0 + 1.2*log2(tail(mixtureA,64)$MXA) + rnorm(64,0,1))
```

Sequins are expected to correlate linearly with the measured expression. Indeed, the measured variable is strongly correlated with the concentration:

```{r, message=FALSE, results='hide', fig.align='center'}
data1 <- createAnaquinData(names=row.names(mixtureA), input=log2(mixtureA$MXA), measured=sim1)
plotScatter(data1, title='Isoform expression (Good)', xlab='Input concentration (log2)', ylab='Measured FPKM (log2)')
```

In our second experiment, the weakly expressed isoforms exhibt stochastic behavior and are clearly not linear with the concentration. Furthermore, there is a limit of quantification (LOQ); below which accuracy of the experiement becomes questionable.

```{r, message=FALSE, results='hide', fig.align='center'}
data2 <- createAnaquinData(names=row.names(mixtureA), input=log2(mixtureA$MXA), measured=sim2)
plotScatter(data2, title='Simulated experiment (Bad)', xlab='Input concentration (log2)', ylab='Measured FPKM (log2)')
```

# Quantifying transcriptome assembly

To estimate the RNA-Seq assembly transcript, we need to run a transcriptome assember, i.e., a software that can assemble transcripts and estimates their abundances. Our workflow guide has the details.

Here, we use a data set estimated by Cuffcompare. Load the data set:

```{r}
data('seqCuffcompare')
head(seqCuffcompare)
```

The first column gives the input concentration for each sequin in attomol/ul. The second column is the measured sensitivity by Cufflinks. Run the following R-code to generate an assembly plot.

```{r, message=FALSE, results='hide', fig.align='center'}
title <- 'Assembly Plot'
xlab  <- 'Input Concentration (log2)'
ylab  <- 'Sensitivity'
anaquin <- createAnaquinData(names=row.names(seqCuffcompare), input=log2(seqCuffcompare$InputConcent), sensitivity=seqCuffcompare$Sn)
plotSensitivity(anaquin, title=title, xlab=xlab, ylab=ylab, showLOA=TRUE)
```

The fitted logistic curve revels clear relationship between input concentration and sensitivity. Unsurprisingly, Cufflinks is more sensitive to highly expressed isoforms. The limit-of-assembly (LOA) is defined as the intersection of the curve to sensitivity of 0.70.

# Quantifying gene expression

Quantifying gene/isoform expression involves building a linear model between input concentration and measured FPKM. In this section, we consider a dataset estimated by Cufflinks.

Load the data set:

```{r}
data('seqCufflinks')
head(seqCufflinks)
```

The first column gives the input concentration for each sequin in attomol/ul. The other columns are the FPKM values for each replicate (three replicates in total). The following code will quantify the the first replicate:

```{r, message=FALSE, fig.align='center'}
title <- 'Gene Expression'
xlab  <- 'Input Concentration (log2)'
ylab  <- 'FPKM (log2)'
anaquin <- createAnaquinData(names=row.names(seqCufflinks), input=log2(seqCufflinks$InputConcent), measured=log2(seqCufflinks$Observed1))
plotScatter(anaquin, title=title, xlab=xlab, ylab=ylab, showLOQ=TRUE)
```

Coefficient of determination is over 0.90; over 90% of the variation (eg: technical bias) can be explained by the model. LOQ is 3.78 attomol/ul, this is the estimated emphirical detection limit.

We can also quantify multiple replicates:

```{r, message=FALSE, fig.align='center'}
title <- 'Gene Expression'
xlab  <- 'Input Concentration (log2)'
ylab  <- 'FPKM (log2)'
anaquin <- createAnaquinData(names=row.names(seqCufflinks), input=log2(seqCufflinks$InputConcent), measured=log2(seqCufflinks[,2:4]))
plotScatter(anaquin, title=title, xlab=xlab, ylab=ylab, showLOQ=TRUE)
```

# Differential analysis

In this section, we show how to quantify differential expression analysis between expected fold-change to measured fold-change. We apply our method to a data set comparing K562 to GM12878 [1]. Load the dataset:

```{r}
data('seqDESeq2')
head(seqDESeq2)
```

For each of the sequin gene, we have expected log-fold change, measured log-fold change, standard deviation, p-value, q-value and mean. The estimation was done by DESeq2.

Run the following code to construct a folding plot:

```{r, fig.align='center'}
title <- 'Gene Fold Change'
xlab  <- 'Expected fold change (log2)'
ylab  <- 'Measured fold change (log2)'
anaquin <- createAnaquinData(names=row.names(seqDESeq2), expected=seqDESeq2$ExpLFC, measured=seqDESeq2$ObsLFC)
plotScatter(anaquin, title=title, xlab=xlab, ylab=ylab, showAxis=TRUE, showLOQ=FALSE)
```

Outliers are obvious throughout the reference scale. Overall, DESeq2 is able to account for 78% of the variation.

We can also construct a ROC plot. Check [1] for how the true-positives and false-positives are defined.

```{r, fig.align='center'}
title <- 'ROC Plot'
anaquin <- createAnaquinData(names=row.names(seqDESeq2), ratio=seqDESeq2$ExpLFC, measured=seqDESeq2$ObsLFC, score=1-seqDESeq2$Pval, label=seqDESeq2$Label)
plotROC(anaquin, title=title, refRats=0)
```

Furthermore, we can plot limit of detection ratio (LODR) curves:

```{r, fig.align='center'}
xlab  <- 'Average Counts'
ylab  <- 'P-value'
title <- 'LODR Curves'
anaquin <- createAnaquinData(names=row.names(seqDESeq2), measured=seqDESeq2$Mean, ratio=seqDESeq2$ExpLFC, pval=seqDESeq2$Pval, qval=seqDESeq2$Qval)
plotLODR(anaquin, xlab=xlab, ylab=ylab, title=title, FDR=0.1, showConf=TRUE)
```
