#
#  Copyright (C) 2015 - Garvan Institute of Medical Research
#
#  Ted Wong, Bioinformatic Software Engineer at Garvan Institute
#

.classify <- function(data, logFC=0, p=0.1)
{
    #
    # Anything that has a fold-change more than logFC is expected to be differential expressed.
    # The following scenarios are possible:
    #
    #    TP: fold-change with more than logFC and expressed
    #    FP: fold-change with at most logFC and expressed
    #    FN: fold-change with more than logFC and not expressed
    #    TN: fold-change with at most logFC and not expressed
    #
    
    for (id in rownames(data))
    {
        if  (is.na(data[id,]$pval) || is.nan(data[id,]$pval))
        {
            data[id,]$class <- NA
        }
        else if (data[id,]$pval <= p)   # Differential expressed
        {
            data[id,]$class <- ifelse(abs(data[id,]$known) <= logFC, 'FP', 'TP')
        }
        else                         # Non-differential expressed
        {
            data[id,]$class <- ifelse(abs(data[id,]$known) <= logFC, 'TN', 'FN')
        }
    }
    
    print(sprintf("Detected %d false positives", nrow(data[data$class=='FP',])))
    print(sprintf("Detected %d true positives",  nrow(data[data$class=='TP',])))
    print(sprintf("Detected %d true negatives",  nrow(data[data$class=='TN',])))
    print(sprintf("Detected %d false negatives", nrow(data[data$class=='FN',])))
    
    data
}

#
# Analyze a result object generated by DESeq2
#

.analyzeDESeq2 <- function(r, mix, p=0.1, logFC=0)
{
    # List of known sequins
    known <- as.character(row.names(mix$genes))
    
    # Genes that have been detected in the experiment    
    detected <- rownames(r) %in% known
    
    # Filter out to only the rows with sequins
    r <- r[detected,]
    
    print(sprintf("Detected %d known sequins", length(known)))
    print(sprintf("Detected %d experimental genes", length(rownames(r))))
    print(sprintf("%d sequins failed to detect", length(known) - length(rownames(r))))
    
    #
    # Create a data-frame for all sequins defined, whether it's detected. The fold-changes are
    # on the logarithmic scale.
    #
    # In this context, NaN refers to undetected sequins while NA refers to detected
    # but untested sequins
    #
    
    d <- data.frame(known=rep(NaN, length(known)),
                    measured=rep(NaN, length(known)),
                    pval=rep(NaN, length(known)),
                    expressed=rep(NaN, length(known)),
                    class=rep(NaN, length(known)))
    rownames(d) <- known
    
    for (id in known)
    {
        d[id,]$known <- mix$genes[row.names(mix$genes)==id,]$logFold
    }
    
    for (id in rownames(r))
    {
        d[id,]$pval      <- r[id,]$padj
        d[id,]$measured  <- r[id,]$log2FoldChange
        d[id,]$expressed <- ifelse(r[id,]$padj <= p, 'T', 'F')
    }
    
    # Sort by adjusted p-values so that the expressed genes are at the front
    d <- d[with(d, order(pval)),]
    
    # Fit a simple-linear regression model on the filtered sequins
    m <- lm(known ~ measured, d[is.finite(d$measured),])
    
    r     <- cor(as.numeric(d$known), as.numeric(d$measured))
    r2    <- summary(m)$r.squared
    slope <- coef(m)["known"]
    
    d <- .classify(d, logFC)

    p <- d[!is.na(d$pval),]
    p <- d[!is.nan(d$pval),]

    plotScatter(p$known, p$measured, row.names(p), isLog=TRUE)
    
    r <- list(data=d, r=r, r2=r2, slope=slope, m=m)
    class(r) <- c("TransDiff")
    r
}
