---
title: "TransQuin Report"
date: "30 November 2015"
author: "Ted Wong (t.wong@garvan.org.au)"
output: pdf_document
---

\fontsize{13pt}{13pt}\selectfont

\setlength{\parskip}{13.0pt}

# Introduction

An accurate quantification of variability in RNA sequencing analysis is critical for bioinformatics analysis. To control for these
sources of variabiltiy, [Garvan Institute of Medical Research](http://www.garvan.org.au) has developed a set of spiked-in synthetic
sequins known as TransQuin. The sequins emulate a diverse range of genomic features and they are combined together across a range of concentration to formulate a mixture to emulate quantitative features.

Anaquin is the statistical software toolkit built for TransQuin. It is designed to analyze outputs generated by the NGS tools, 
and accommodates the analysis with detailed statistics on the sequins. Please refer to [www.sequin.xyz](www.sequin.xyz) for
more details.

This report was generated by the following Anaquin command:

```python
anaquin -r PDF -t TransAlign -rgtf ATR001.v032.gtf -m MTR002.v013.csv ...
```

# Inputs

The following reference files were used:

* ATR001.v032.gtf

The following mixtures were used:

* MTR002.v013.csv

\newpage

\begingroup
    \fontsize{20pt}{20pt}\selectfont
        __Contents__
\endgroup

1.  RNA Alignment
    1.  Summary Statistics
    2.  ROC curve
    3.  Sequins
\setlength\itemsep{1em}
2.  RNA Assembly
    1.  Summary Statistics
    2.  Abundance plot
    3.  Sequins

3. Apprendix:
    1.  Mixtures

\newpage

\begingroup
    \fontsize{20pt}{20pt}\selectfont
        __RNA Alignment__
\endgroup

# 1. Summary Statistics

```python
Summary for dataset: sampled.sam

   Unmapped:   0 reads
   Experiment: 285619 reads
   Synthetic:  0 reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      468524 exons
   Query:      179748 introns
   Query:      160672 bases

   Dilution:   0

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by the performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by the performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.999991
   Accuarcy:    0.990056
   Detection:   0.472069 (R1_91)

   -------------------- Intron level --------------------

   Sensitivity: 0.999894
   Accuarcy:    0.942269
   Detection:   0.118017 (R1_33)

   -------------------- Base level --------------------

   Sensitivity: 0.443994
   Accuarcy:    0.602843
   Detection:   0.472069 (R1_91)
```

# 2. ROC Curve

```{r echo=FALSE, message=FALSE, cache=TRUE}
library(ROCR)

#par(mfrow= c(2,2))

data(ROCR.xval)

pred <- prediction(ROCR.xval$predictions, ROCR.xval$labels)
perf <- performance(pred,"tpr","fpr")

plot(perf)
```

\newpage

# Appendix: Mixtures

```{r}
library(Anaquin)

m <- loadMixture()
m$isoforms
```

\newpage




You can also embed plots, for example:

```{r, echo=FALSE}
plot(cars)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
