\documentclass[]{article}
\title{\textbf{Sequin Report}}

\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

\date{}

\begin{document}
\maketitle



10/11/2017 12:53 PM

\textbf{INPUT FILES{~}}

\textbf{}\\

\textbf{USER-SUPPLIED FILES}

User sequin variants:{ }{ }{ }/path/to/L.V.97.sequin.raw\_variants.vcf{
}{ }{[}user-supplied file{]}

User sample variants:{ }{ }{ }/path/to/L.V.97.human.raw\_variants.vcf{
}{ }{[}user-supplied file{]}

Excluded edge size:{ }{ }{ }{ }600bp{ }{ }{ }{ }{ }{[}default value{]}

User regions (hg38):{ }{ }{ }NONE

\textbf{REFERENCE FILES}

Reference sequin variants:{ }{ }{ }/path/to/sequins\_variants.hg38.vcf{
}{ }{[}anaquin pre-packed file{]}

Sequin regions (hg38):{ } { }{
}/path/to/sequin\_regions.combined.hg38.bed{ }{[}anaquin pre-packed
file{]}

Comparison regions:{ }{ }{ }/path/to/comparison\_regions.bed{ }{
}{[}anaquin output file{]}

-----

\textbf{ORIGNAL ALIGNMENT}

Sample:{ }{ }{ }{ }{ }930900193 (99.38\%)

Sequin:{ }{ }{ }{ }{ }3197725 (0.34\%)

Dilution:{ }{ }{ }{ }{ }1\%

Unmapped:{ }{ }{ }{ }2561895 (0.27\%)

\textbf{READ FLIPPING}

Flipped pairs:{ }{ }{ }{ }58475 (61.58\%)

Ambiguous pairs:{ }{ }{ }{ }3983 (4.19\%)

Hanging reads:{ }{ }{ }{ } 0 (0.00\%)

\textbf{CALIBRATION}

\emph{Before calibration:}

Mean sample coverage:{ }{ }{ }42.69x

Mean sequin coverage:{ }{ }{ }645.23x

\emph{After calibration:}

Scaling Factor:{ }{ }{ }{ }0.134112 Â± 0.124793

Mean sequin coverage:{ }{ }{ }41.84x

\textbf{VISUALISATION}

\textbf{}\\

\textbf{Use the following files for visualization in IGV}

Sample alignments:{ }{ }{ }{ }/path/to/sample\_alignments.bam{ }

Index file:{ }{ }{ }{ }{ }/path/to/sample\_alignments.bam.bai

Sample variants:{ }{ }{ }{ }/path/to/sample\_variants.vcf

Sequin alignments:{ }{ }{ }{ }/path/to/sequin\_alignments.bam{ }

Sequin variants:{ }{ }{ }{ }/path/to/sequin\_variants.vcf

Index file:{ }{ }{ }{ }{ }/path/to/sequin\_alignments.bam.bai{ }

Sequin regions:{ }{ }{ }{ }/path/to/sequin\_regions.bed{ }{ }

\textbf{SEQUENCING ERROR}

\textbf{}\\

\textbf{GENOME COMPARISONS}

\textbf{}

\textbf{}\\

\textbf{GERMLINE VARIANTS}

\textbf{}\\

\textbf{ALL VARIANTS}:{ }{ }{ }{
}{[}CS,CI,GS,GI,LoGC,VLGC,HiGC,VHGC,MS,SH,LH,SD,LD,ST,LT,SQ,LQ{]}

Comparison regions:{ }{ }{ }270

Comparison size (bp):{ }{ }{ }157134

{ }{ }{ }{ }{ }\textbf{Total{ }SNVs{ }INDELs}

Reference sequin variants:{ }{ }{ }170{ }107{ }63

True positives (TP):{ }{ }{ }{ }152{ }101{ }51

False positives (FP):{ }{ }{ }21{ }12{ }9

False negatives (FN):{ }{ }{ }21{ }12{ }9

Sensitivity:{ }{ }{ }{ }0.89{ }0.94{ }0.81

Precision{ }{ }{ }{ }{ }0.873

F1 Score{ }{ }{ }{ }{ }0.883

FDR Rate{ }{ }{ }{ }{ }0.126

FPR (FP/kb):{ }{ }{ }{ }0.13{ }0.08{ }0.05

{ }{ }{ }{ }{ }\textbf{COV\_Median{ }QUAL\_Median{ }QUAL\_10\%{
}QUAL\_90\%}

TP:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

FP:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

FN:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

\textbf{COMMON VARIANTS}{ }{ }{ }{[}GS,GI standards{]}

Comparison regions:{ }{ }{ }102

Comparison space (bp):{ }{ }{ }61050

{ }{ }{ }{ }{ }\textbf{Total{ }SNVs{ }INDELs{ }HOM{ }HET}

Reference sequin variants:{ }{ }{ }102{ }87{ }15{ }X{ }X

True positives:{ }{ }{ }{ }97{ }82{ }15{ }X{ }X

False positives:{ }{ }{ }{ }11{ }7{ }4{ }X{ }X

False negatives (FN):{ }{ }{ }21{ }12{ }9

Sensitivity:{ }{ }{ }{ }0.95{ }0.94{ }1.00{ }X{ }X

Precision{ }{ }{ }{ }{ }0.873

F1 Score{ }{ }{ }{ }{ }0.883

FDR Rate{ }{ }{ }{ }{ }0.126

FPR (FP/kb):{ }{ }{ }{ }0.18{ }0.11{ }0.07{ }X{ }X

{ }{ }{ }{ }{ }\textbf{COV\_Median{ }QUAL\_Median{ }QUAL\_10\%{
}QUAL\_90\%}

TP:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

FP:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

FN:{ }{ }{ }{ }{ }X{ }{ }x{ }{ }x{ }{ }x

\textbf{GC POOR/RICH}{ }{ }{ }{ }{[}LoGC,VLGC,HiGC,VHGC standards{]}

Comparison regions:{ }{ }{ }20

Comparisons space (bp):{ }{ }{ }11058 bp

Comparison variants:{ }{ }{ }20 SNVs

Sensitivity:{ }{ }{ }{ }0.95

FPR (FP/kb):{ }{ }{ }{ }0.27

{ }{ }{ }{ }{ }\textbf{TP{ }FN{ }Sn{ }FP{ }FPR(FP/kb)}

GC poor: { }{ }{ }{ }{ }10{ }0 { }1.00{ }0{ }0.00

GC rich:{ }{ }{ }{ }{ }9{ }1 { }0.90{ }3{ }0.54

\textbf{SIMPLE REPEATS}{ }{ }{ }{ }{[}MS,SH,LH,SD,LD,ST,LT,SQ,LQ
standards{]}

Comparison regions:{ }{ }{ }53

Comparison space (bp):{ }{ }{ }28026

Comparison variants:{ }{ }{ }48 INDELs

Sensitivity:{ }{ }{ }{ }0.75

FPR (FP/kb):{ }{ }{ }{ }0.21

\textbf{{ }{ }{ }{ }{ }Short{ }Long}

Homo:{ }{ }{ }{ }{ }6/6{ }1/6

Di:{ }{ }{ }{ }{ }6/6{ }3/6

Tri: { }{ }{ }{ }{ }5/6{ }5/6

Quad:{ }{ }{ }{ }{ }5/6{ }5/6

Total:{ }{ }{ }{ }{ }22/24{ }14/24

\textbf{}

\textbf{SOMATIC VARIANTS}

\textbf{}\\

\textbf{SOMATIC VARIANTS}

Comparison regions:{ }{ }{ }102

Comparison space (bp):{ }{ }{ }61050

{ }{ }{ }{ }{ }\textbf{Total{ }SNVs{ }INDELs{ }HOM{ }HET}

Reference sequin variants:{ }{ }{ }102{ }87{ }15{ }X{ }X

True positives:{ }{ }{ }{ }97{ }82{ }15{ }X{ }X

False negatives (FN):{ }{ }{ }21{ }12{ }9

Sensitivity:{ }{ }{ }{ }0.95{ }0.94{ }1.00{ }X{ }X

Detection Sensitivity:{ }{ }{ }0.0036875 (attomol/ul){~}

\textbf{Linear regression (allele frequency) (log2 scale)}

Slope:{ }{ }{ }{ }{ }1.10

Correlation:{ }{ }{ }{ }0.96

R2:{ }{ }{ }{ }{ }0.93

P-value:{~ ~ { }{ }{ }{ }}0.00

\textbf{AUC values}

XXX:

\textbf{MICROSATELLITES}{ }{ }{ }{[}MS,SH,LH,SD,LD,ST,LT,SQ,LQ
standards{]}

Comparison regions:{ }{ }{ }53

Comparison space (bp):{ }{ }{ }28026

Comparison variants:{ }{ }{ }48 INDELs

Sensitivity:{ }{ }{ }{ }0.75

FPR (FP/kb):{ }{ }{ }{ }0.21

\textbf{CANCER GENES}

\textbf{}

\textbf{INDIVIDUAL SEQUINS}

\textbf{}\\

\textbf{COMMON VARIANTS}

\textbf{Name{ }Chrom:Position{ }Label{ }ReadR{ }ReadV{ }Depth{ }ExpFreq{
}ObsFreq{ }Qual{ }Mutation{ }Visualisation}

GS\_084{ }chr8:5714529{ }TP{ }26{ }20{ }46{ }0.5{ }0.5{ }618.77{ }SNP{
}{[}IGV link{]}

GS\_087{ }chr3:7076599{ }TP{ }23{ }20{ }43{ }0.5{ }0.5{ }712.77{ }SNP{
}{[}IGV link{]}

GS\_036{ }chr18:7456095{ }TP{ }15{ }26{ }41{ }0.5{ }0.5{ }921.77{ }SNP{
}{[}IGV link{]}

GS\_077{ }chr1:9062342{ }TP{ }13{ }17{ }30{ }0.5{ }0.5{ }622.77{ }SNP{
}{[}IGV link{]}

GS\_101{ }chr2:10000522{ }TP{ }30{ }12{ }42{ }0.5{ }0.5{ }360.77{ }SNP{
}{[}IGV link{]}

GS\_068{ }chr10:10023474{ }TP{ }17{ }41{ }58{ }0.5{ }0.5{ }1497.77{
}SNP{ }{[}IGV link{]}

GS\_063{ }chr17:13924116{ }TP{ }25{ }16{ }41{ }0.5{ }0.5{ }518.77{ }SNP{
}{[}IGV link{]}

GS\_054{ }chr19:14694467{ }TP{ }12{ }20{ }32{ }0.5{ }0.5{ }718.77{ }SNP{
}{[}IGV link{]}

GS\_012{ }chr2:15061668{ }TP{ }0{ }45{ }46{ }1{ }1{ }1777.77{ }SNP{
}{[}IGV link{]}

GS\_075{ }chr5:19329360{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{ }SNP{ }{[}IGV
link{]}

GS\_066{ }chr6:20383417{ }TP{ }14{ }22{ }36{ }0.5{ }0.5{ }787.77{ }SNP{
}{[}IGV link{]}

GS\_061{ }chr4:21025552{ }TP{ }9{ }3{ }12{ }0.5{ }0.5{ }45.77{ }SNP{
}{[}IGV link{]}

GS\_027{ }chr2:21386284{ }TP{ }0{ }39{ }39{ }1{ }1{ }1553.77{ }SNP{
}{[}IGV link{]}

GS\_034{ }chr1:24300909{ }TP{ }0{ }48{ }48{ }1{ }1{ }1951.77{ }SNP{
}{[}IGV link{]}

GS\_060{ }chr22:24759176{ }TP{ }27{ }22{ }49{ }0.5{ }0.5{ }727.77{ }SNP{
}{[}IGV link{]}

GS\_081{ }chr11:24894129{ }TP{ }18{ }19{ }37{ }0.5{ }0.5{ }614.77{ }SNP{
}{[}IGV link{]}

GS\_040{ }chr10:25478591{ }TP{ }19{ }23{ }42{ }0.5{ }0.5{ }769.77{ }SNP{
}{[}IGV link{]}

GS\_031{ }chr15:27986646{ }FN{ }-{ }-{ }-{ }1{ }-{ }-{ }SNP{ }{[}IGV
link{]}

GS\_082{ }chr4:31587942{ }TP{ }26{ }13{ }39{ }0.5{ }0.5{ }409.77{ }SNP{
}{[}IGV link{]}

GS\_019{ }chr11:31667833{ }TP{ }0{ }54{ }54{ }1{ }1{ }2174.77{ }SNP{
}{[}IGV link{]}

GS\_088{ }chr9:33425148{ }TP{ }16{ }16{ }32{ }0.5{ }0.5{ }547.77{ }SNP{
}{[}IGV link{]}

GS\_020{ }chr4:34759897{ }FN{ }-{ }-{ }-{ }1{ }-{ }-{ }SNP{ }{[}IGV
link{]}

GS\_010{ }chr2:37662839{ }TP{ }0{ }40{ }40{ }1{ }1{ }1736.77{ }SNP{
}{[}IGV link{]}

GS\_029{ }chr18:37720536{ }TP{ }0{ }49{ }49{ }1{ }1{ }2003.77{ }SNP{
}{[}IGV link{]}

GS\_026{ }chr12:38017759{ }TP{ }0{ }47{ }47{ }1{ }1{ }1993.77{ }SNP{
}{[}IGV link{]}

GS\_059{ }chr2:38058630{ }TP{ }24{ }25{ }49{ }0.5{ }0.5{ }888.77{ }SNP{
}{[}IGV link{]}

GS\_058{ }chr2:42277269{ }TP{ }37{ }11{ }49{ }0.5{ }0.5{ }294.77{ }SNP{
}{[}IGV link{]}

GS\_100{ }chr11:42416907{ }TP{ }17{ }26{ }43{ }0.5{ }0.5{ }872.77{ }SNP{
}{[}IGV link{]}

GS\_042{ }chr6:42684928{ }TP{ }24{ }12{ }36{ }0.5{ }0.5{ }403.77{ }SNP{
}{[}IGV link{]}

GS\_007{ }chr9:42927778{ }TP{ }0{ }45{ }45{ }1{ }1{ }1892.77{ }SNP{
}{[}IGV link{]}

GS\_032{ }chr5:43535888{ }TP{ }0{ }41{ }41{ }1{ }1{ }1769.77{ }SNP{
}{[}IGV link{]}

GS\_009{ }chr3:45647980{ }TP{ }0{ }37{ }37{ }1{ }1{ }1517.77{ }SNP{
}{[}IGV link{]}

GS\_038{ }chr14:50087682{ }TP{ }23{ }15{ }38{ }0.5{ }0.5{ }509.77{ }SNP{
}{[}IGV link{]}

GS\_064{ }chr19:50219912{ }TP{ }17{ }29{ }46{ }0.5{ }0.5{ }1081.77{
}SNP{ }{[}IGV link{]}

GS\_098{ }chr5:53809384{ }TP{ }23{ }19{ }43{ }0.5{ }0.5{ }633.77{ }SNP{
}{[}IGV link{]}

GS\_083{ }chr16:54487171{ }TP{ }21{ }23{ }44{ }0.5{ }0.5{ }849.77{ }SNP{
}{[}IGV link{]}

GI\_044{ }chr11:54648345{ }TP{ }31{ }19{ }51{ }0.5{ }0.5{ }575.73{
}Deletion{ }{[}IGV link{]}

GS\_073{ }chr15:57594555{ }TP{ }22{ }23{ }45{ }0.5{ }0.5{ }842.77{ }SNP{
}{[}IGV link{]}

GI\_086{ }chr1:58701655{ }TP{ }27{ }13{ }44{ }0.5{ }0.5{ }419.73{
}Deletion{ }{[}IGV link{]}

GI\_048{ }chr1:60005803{ }TP{ }15{ }8{ }34{ }0.5{ }0.5{ }173.73{
}Insertion{ }{[}IGV link{]}

GS\_096{ }chr3:61787430{ }TP{ }28{ }21{ }49{ }0.5{ }0.5{ }741.77{ }SNP{
}{[}IGV link{]}

GS\_053{ }chr5:67588277{ }TP{ }19{ }27{ }46{ }0.5{ }0.5{ }899.77{ }SNP{
}{[}IGV link{]}

GS\_091{ }chr12:71427517{ }TP{ }21{ }19{ }40{ }0.5{ }0.5{ }647.77{ }SNP{
}{[}IGV link{]}

GS\_095{ }chr13:71948318{ }TP{ }31{ }19{ }50{ }0.5{ }0.5{ }636.77{ }SNP{
}{[}IGV link{]}

GS\_011{ }chr8:72748366{ }TP{ }0{ }47{ }47{ }1{ }1{ }1990.77{ }SNP{
}{[}IGV link{]}

GS\_047{ }chr7:79105367{ }TP{ }20{ }25{ }45{ }0.5{ }0.5{ }834.77{ }SNP{
}{[}IGV link{]}

GI\_085{ }chr16:82894695{ }TP{ }9{ }15{ }25{ }0.5{ }0.5{ }357.73{
}Deletion{ }{[}IGV link{]}

GI\_094{ }chr2:82976852{ }TP{ }15{ }24{ }48{ }0.5{ }0.5{ }389.73{
}Insertion{ }{[}IGV link{]}

GI\_006{ }chr5:84326515{ }TP{ }0{ }32{ }33{ }1{ }1{ }1402.73{ }Deletion{
}{[}IGV link{]}

GS\_018{ }chr11:85070085{ }TP{ }0{ }33{ }33{ }1{ }1{ }1328.77{ }SNP{
}{[}IGV link{]}

GS\_004{ }chr9:85568589{ }TP{ }0{ }40{ }40{ }1{ }1{ }1555.77{ }SNP{
}{[}IGV link{]}

GS\_003{ }chr2:85940422{ }TP{ }0{ }45{ }45{ }1{ }1{ }1867.77{ }SNP{
}{[}IGV link{]}

GS\_056{ }chr14:86093328{ }TP{ }29{ }22{ }51{ }0.5{ }0.5{ }727.77{ }SNP{
}{[}IGV link{]}

GS\_078{ }chr11:86212150{ }TP{ }23{ }15{ }38{ }0.5{ }0.5{ }509.77{ }SNP{
}{[}IGV link{]}

GI\_074{ }chr5:86492393{ }TP{ }27{ }20{ }51{ }0.5{ }0.5{ }430.73{
}Insertion{ }{[}IGV link{]}

GS\_001{ }chr13:88107779{ }TP{ }0{ }47{ }47{ }1{ }1{ }1948.77{ }SNP{
}{[}IGV link{]}

GS\_013{ }chr8:89181017{ }TP{ }0{ }35{ }35{ }1{ }1{ }1460.77{ }SNP{
}{[}IGV link{]}

GS\_041{ }chr6:89296205{ }TP{ }25{ }15{ }40{ }0.5{ }0.5{ }466.77{ }SNP{
}{[}IGV link{]}

GS\_090{ }chr16:89739568{ }TP{ }17{ }27{ }44{ }0.5{ }0.5{ }844.77{ }SNP{
}{[}IGV link{]}

GS\_039{ }chr15:92824265{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{ }SNP{ }{[}IGV
link{]}

GS\_037{ }chr9:93265888{ }TP{ }24{ }12{ }36{ }0.5{ }0.5{ }365.77{ }SNP{
}{[}IGV link{]}

GS\_079{ }chr15:93602662{ }TP{ }25{ }12{ }37{ }0.5{ }0.5{ }367.77{ }SNP{
}{[}IGV link{]}

GS\_089{ }chr9:95332055{ }TP{ }20{ }22{ }42{ }0.5{ }0.5{ }740.77{ }SNP{
}{[}IGV link{]}

GI\_005{ }chrX:98156963{ }TP{ }0{ }43{ }43{ }1{ }1{ }1570.73{
}Insertion{ }{[}IGV link{]}

GS\_102{ }chr8:99091055{ }TP{ }13{ }22{ }35{ }0.5{ }0.5{ }764.77{ }SNP{
}{[}IGV link{]}

GS\_080{ }chr7:101692183{ }TP{ }13{ }12{ }25{ }0.5{ }0.5{ }390.77{ }SNP{
}{[}IGV link{]}

GS\_067{ }chr6:102283832{ }TP{ }26{ }15{ }41{ }0.5{ }0.5{ }430.77{ }SNP{
}{[}IGV link{]}

GS\_050{ }chr7:104663550{ }TP{ }15{ }31{ }46{ }0.5{ }0.5{ }1056.77{
}SNP{ }{[}IGV link{]}

GS\_076{ }chr1:105421342{ }TP{ }23{ }10{ }33{ }0.5{ }0.5{ }286.77{ }SNP{
}{[}IGV link{]}

GS\_052{ }chr12:107380095{ }TP{ }29{ }23{ }53{ }0.5{ }0.5{ }715.77{
}SNP{ }{[}IGV link{]}

GS\_025{ }chr13:107675604{ }TP{ }0{ }53{ }53{ }1{ }1{ }2200.77{ }SNP{
}{[}IGV link{]}

GS\_071{ }chr13:109015195{ }TP{ }37{ }13{ }50{ }0.5{ }0.5{ }398.77{
}SNP{ }{[}IGV link{]}

GS\_024{ }chr2:110013774{ }TP{ }0{ }24{ }24{ }1{ }1{ }952.77{ }SNP{
}{[}IGV link{]}

GS\_046{ }chr7:113042503{ }TP{ }20{ }16{ }36{ }0.5{ }0.5{ }553.77{ }SNP{
}{[}IGV link{]}

GS\_022{ }chr3:115437301{ }TP{ }0{ }45{ }45{ }1{ }1{ }1881.77{ }SNP{
}{[}IGV link{]}

GS\_028{ }chr4:117274500{ }TP{ }0{ }41{ }41{ }1{ }1{ }1712.77{ }SNP{
}{[}IGV link{]}

GI\_099{ }chr11:124829273{ }TP{ }30{ }23{ }54{ }0.5{ }0.5{ }837.73{
}Insertion{ }{[}IGV link{]}

GI\_045{ }chr11:132015632{ }TP{ }14{ }13{ }40{ }0.5{ }0.5{ }465.73{
}Insertion{ }{[}IGV link{]}

GS\_017{ }chr4:134486115{ }TP{ }0{ }49{ }49{ }1{ }1{ }2086.77{ }SNP{
}{[}IGV link{]}

GS\_055{ }chr8:138586475{ }TP{ }15{ }18{ }33{ }0.5{ }0.5{ }652.77{ }SNP{
}{[}IGV link{]}

GS\_033{ }chr2:141264222{ }TP{ }0{ }50{ }50{ }1{ }1{ }2145.77{ }SNP{
}{[}IGV link{]}

GS\_051{ }chr4:145380496{ }TP{ }25{ }25{ }50{ }0.5{ }0.5{ }863.77{ }SNP{
}{[}IGV link{]}

GS\_065{ }chr5:149421815{ }TP{ }14{ }21{ }35{ }0.5{ }0.5{ }756.77{ }SNP{
}{[}IGV link{]}

GI\_093{ }chr7:150556565{ }TP{ }20{ }12{ }36{ }0.5{ }0.5{ }330.73{
}Deletion{ }{[}IGV link{]}

GS\_057{ }chr5:153549030{ }TP{ }30{ }8{ }38{ }0.5{ }0.5{ }210.77{ }SNP{
}{[}IGV link{]}

GS\_023{ }chr2:155294875{ }TP{ }0{ }53{ }53{ }1{ }1{ }2127.77{ }SNP{
}{[}IGV link{]}

GS\_014{ }chr6:159450081{ }TP{ }0{ }38{ }38{ }1{ }1{ }1619.77{ }SNP{
}{[}IGV link{]}

GS\_008{ }chr3:160029898{ }TP{ }0{ }40{ }41{ }1{ }1{ }1607.77{ }SNP{
}{[}IGV link{]}

GI\_030{ }chr1:160205419{ }TP{ }0{ }34{ }41{ }1{ }1{ }1457.73{
}Insertion{ }{[}IGV link{]}

GS\_002{ }chr4:160613491{ }TP{ }0{ }41{ }41{ }1{ }1{ }1726.77{ }SNP{
}{[}IGV link{]}

GS\_021{ }chr5:161927178{ }TP{ }0{ }34{ }34{ }1{ }1{ }1429.77{ }SNP{
}{[}IGV link{]}

GI\_070{ }chr4:167080750{ }TP{ }23{ }13{ }37{ }0.5{ }0.5{ }430.73{
}Deletion{ }{[}IGV link{]}

GI\_049{ }chr3:173057114{ }TP{ }8{ }23{ }33{ }0.5{ }0.5{ }395.73{
}Insertion{ }{[}IGV link{]}

GS\_043{ }chr3:173135148{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{ }SNP{ }{[}IGV
link{]}

GS\_035{ }chr3:189655569{ }TP{ }0{ }49{ }49{ }1{ }1{ }1999.77{ }SNP{
}{[}IGV link{]}

GS\_069{ }chr3:194241602{ }TP{ }12{ }45{ }57{ }0.5{ }0.5{ }1719.77{
}SNP{ }{[}IGV link{]}

GI\_097{ }chr1:206094168{ }TP{ }22{ }21{ }50{ }0.5{ }0.5{ }779.73{
}Deletion{ }{[}IGV link{]}

GS\_092{ }chr2:214030136{ }TP{ }27{ }18{ }45{ }0.5{ }0.5{ }611.77{ }SNP{
}{[}IGV link{]}

GS\_072{ }chr2:228157894{ }TP{ }21{ }19{ }40{ }0.5{ }0.5{ }664.77{ }SNP{
}{[}IGV link{]}

GS\_015{ }chr1:230128871{ }TP{ }0{ }36{ }36{ }1{ }1{ }1414.77{ }SNP{
}{[}IGV link{]}

GS\_062{ }chr1:238551365{ }TP{ }37{ }9{ }46{ }0.5{ }0.5{ }198.77{ }SNP{
}{[}IGV link{]}

GS\_016{ }chr1:248001905{ }TP{ }0{ }41{ }41{ }1{ }1{ }1673.77{ }SNP{
}{[}IGV link{]}

\textbf{DIFFICULT VARIANTS}

\textbf{Name{ }Chrom:Position{ }Label{ }ReadR{ }ReadV{ }Depth{ }ExpFreq{
}ObsFreq{ }Qual{ }Genotype{ }Context{ }Mutation{ }Visualisation}

\textbf{HiGC\_02{ }chr7:1227679{ }TP{ }26{ }18{ }44{ }0.5{ }0.5{
}626.77{ }Heterozygous{ }HighGC{ }SNP{ }{[}IGV link{]}}

\textbf{HiGC\_10{ }chr6:17728632{ }TP{ }23{ }20{ }43{ }0.5{ }0.5{
}699.77{ }Heterozygous{ }HighGC{ }SNP{ }{[}IGV link{]}}

\textbf{HiGC\_09{ }chr14:103539360{ }TP{ }28{ }13{ }41{ }0.5{ }0.5{
}352.77{ }Heterozygous{ }HighGC{ }SNP{ }{[}IGV link{]}}

\textbf{HiGC\_08{ }chr3:146250691{ }TP{ }16{ }28{ }44{ }0.5{ }0.5{
}974.77{ }Heterozygous{ }HighGC{ }SNP{ }{[}IGV link{]}}

\textbf{HiGC\_03{ }chr2:178413882{ }TP{ }31{ }9{ }40{ }0.5{ }0.5{
}250.77{ }Heterozygous{ }HighGC{ }SNP{ }{[}IGV link{]}}

\textbf{LD\_004{ }chr1:90824057{ }TP{ }7{ }18{ }38{ }0.5{ }0.5{ }333.73{
}Heterozygous{ }LongDinucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LD\_001{ }chr8:98185045{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongDinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LD\_100{ }chr10:98852112{ }TP{ }5{ }15{ }33{ }0.5{ }0.5{ }258.7{
}Heterozygous{ }LongDinucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LD\_006{ }chr8:130573543{ }TP{ }6{ }8{ }27{ }0.5{ }0.5{ }118.73{
}Heterozygous{ }LongDinucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LD\_003{ }chr6:150701416{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongDinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LD\_002{ }chr1:231532130{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongDinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LH\_009{ }chr8:9584163{ }TP{ }5{ }13{ }32{ }0.5{ }0.5{ }128.82{
}Heterozygous{ }LongHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{LH\_005{ }chr20:17189849{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{LH\_006{ }chr13:30193453{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{LH\_011{ }chr16:78345638{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{LH\_008{ }chr9:91084893{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{LH\_007{ }chr1:115251304{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{LQ\_007{ }chr20:2015767{ }TP{ }6{ }11{ }46{ }0.5{ }0.5{ }392.73{
}Heterozygous{ }LongQuadNucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LQ\_008{ }chr8:3494553{ }TP{ }5{ }19{ }37{ }0.5{ }0.5{ }707.73{
}Heterozygous{ }LongQuadNucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LQ\_001{ }chr10:6988515{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongQuadNucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LQ\_004{ }chr16:50901756{ }TP{ }13{ }10{ }51{ }0.5{ }0.5{
}341.73{ }Heterozygous{ }LongQuadNucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{LQ\_005{ }chr18:77819648{ }TP{ }17{ }3{ }38{ }0.5{ }0.5{ }38.73{
}Heterozygous{ }LongQuadNucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LQ\_003{ }chr6:140286316{ }TP{ }5{ }7{ }28{ }0.5{ }0.5{ }232.73{
}Heterozygous{ }LongQuadNucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LT\_019{ }chr7:18272912{ }TP{ }13{ }16{ }46{ }0.5{ }0.5{
}462.73{ }Heterozygous{ }LongTrinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{LT\_011{ }chrX:31384308{ }TP{ }6{ }16{ }43{ }0.5{ }0.5{ }520.73{
}Heterozygous{ }LongTrinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LT\_018{ }chr19:35910644{ }TP{ }9{ }7{ }29{ }0.5{ }0.5{ }201.73{
}Heterozygous{ }LongTrinucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{LT\_005{ }chr6:36549328{ }TP{ }10{ }10{ }23{ }0.5{ }0.5{
}275.73{ }Heterozygous{ }LongTrinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{LT\_013{ }chr13:83309637{ }TP{ }15{ }14{ }46{ }0.5{ }0.5{
}421.73{ }Heterozygous{ }LongTrinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{LT\_001{ }chr8:117790868{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }LongTrinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{LoGC\_05{ }chr13:38220157{ }TP{ }30{ }16{ }46{ }0.5{ }0.5{
}533.77{ }Heterozygous{ }LowGC{ }SNP{ }{[}IGV link{]}}

\textbf{LoGC\_04{ }chr8:76538464{ }TP{ }23{ }25{ }48{ }0.5{ }0.5{
}929.77{ }Heterozygous{ }LowGC{ }SNP{ }{[}IGV link{]}}

\textbf{LoGC\_02{ }chr12:126963049{ }TP{ }1{ }29{ }31{ }0.5{ }1{
}1120.77{ }Heterozygous{ }LowGC{ }SNP{ }{[}IGV link{]}}

\textbf{LoGC\_01{ }chr5:129909395{ }TP{ }26{ }22{ }48{ }0.5{ }0.5{
}707.77{ }Heterozygous{ }LowGC{ }SNP{ }{[}IGV link{]}}

\textbf{LoGC\_03{ }chr1:178354550{ }TP{ }29{ }22{ }51{ }0.5{ }0.5{
}774.77{ }Heterozygous{ }LowGC{ }SNP{ }{[}IGV link{]}}

\textbf{SD\_003{ }chr7:28849446{ }TP{ }10{ }21{ }36{ }0.5{ }0.5{
}667.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{SD\_005{ }chr21:32340143{ }TP{ }17{ }31{ }50{ }0.5{ }0.5{
}1077.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{SD\_006{ }chr5:74230403{ }TP{ }27{ }29{ }59{ }0.5{ }0.5{
}914.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{SD\_004{ }chr7:93771338{ }TP{ }11{ }33{ }46{ }0.5{ }0.5{
}1139.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{SD\_001{ }chr4:189266611{ }TP{ }30{ }11{ }44{ }0.5{ }0.5{
}240.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{SD\_002{ }chr1:236471368{ }TP{ }22{ }27{ }50{ }0.5{ }0.5{
}821.73{ }Heterozygous{ }ShortDinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{SH\_004{ }chr16:25209091{ }TP{ }14{ }9{ }28{ }0.5{ }0.5{
}155.73{ }Heterozygous{ }ShortHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{SH\_006{ }chr6:68186858{ }TP{ }23{ }30{ }57{ }0.5{ }0.5{
}733.73{ }Heterozygous{ }ShortHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{SH\_005{ }chr3:72033879{ }TP{ }25{ }15{ }47{ }0.5{ }0.5{
}278.73{ }Heterozygous{ }ShortHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{SH\_003{ }chr9:106379843{ }TP{ }19{ }24{ }46{ }0.5{ }0.5{
}564.73{ }Heterozygous{ }ShortHomopolymer{ }Insertion{ }{[}IGV link{]}}

\textbf{SH\_010{ }chr6:164716060{ }TP{ }10{ }41{ }54{ }0.5{ }0.5{
}1086.73{ }Heterozygous{ }ShortHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{SH\_008{ }chr1:247187985{ }TP{ }11{ }40{ }54{ }0.5{ }0.5{
}1053.73{ }Heterozygous{ }ShortHomopolymer{ }Deletion{ }{[}IGV link{]}}

\textbf{SQ\_001{ }chr22:22515400{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }ShortQuadNucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{SQ\_006{ }chr12:46499027{ }TP{ }13{ }35{ }52{ }0.5{ }0.5{
}1381.73{ }Heterozygous{ }ShortQuadNucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{SQ\_004{ }chr17:61077532{ }TP{ }31{ }8{ }44{ }0.5{ }0.5{
}240.73{ }Heterozygous{ }ShortQuadNucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{SQ\_002{ }chr18:63732184{ }TP{ }26{ }13{ }41{ }0.5{ }0.5{
}429.73{ }Heterozygous{ }ShortQuadNucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{SQ\_009{ }chr4:139081241{ }TP{ }10{ }27{ }39{ }0.5{ }0.5{
}1066.73{ }Heterozygous{ }ShortQuadNucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{SQ\_005{ }chr1:197086071{ }TP{ }27{ }17{ }44{ }0.5{ }0.5{
}589.73{ }Heterozygous{ }ShortQuadNucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{ST\_016{ }chr1:26008441{ }TP{ }18{ }10{ }31{ }0.5{ }0.5{
}321.73{ }Heterozygous{ }ShortTrinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{ST\_007{ }chr5:76897301{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }ShortTrinucleotideRepeat{ }Deletion{ }{[}IGV link{]}}

\textbf{ST\_008{ }chr8:77483036{ }TP{ }23{ }28{ }54{ }0.5{ }0.5{
}1069.73{ }Heterozygous{ }ShortTrinucleotideRepeat{ }Deletion{ }{[}IGV
link{]}}

\textbf{ST\_005{ }chr4:84811428{ }TP{ }14{ }29{ }46{ }0.5{ }0.5{
}1252.73{ }Heterozygous{ }ShortTrinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{ST\_004{ }chr15:90576935{ }TP{ }0{ }36{ }40{ }0.5{ }1{ }1579.73{
}Heterozygous{ }ShortTrinucleotideRepeat{ }Insertion{ }{[}IGV link{]}}

\textbf{ST\_003{ }chr3:171140421{ }TP{ }29{ }15{ }44{ }0.5{ }0.5{
}505.73{ }Heterozygous{ }ShortTrinucleotideRepeat{ }Insertion{ }{[}IGV
link{]}}

\textbf{VHGC\_17{ }chr11:1571667{ }FN{ }-{ }-{ }-{ }0.5{ }-{ }-{
}Heterozygous{ }VeryHighGC{ }SNP{ }{[}IGV link{]}}

\textbf{VHGC\_13{ }chr18:37254221{ }TP{ }19{ }7{ }26{ }0.5{ }0.5{
}182.77{ }Heterozygous{ }VeryHighGC{ }SNP{ }{[}IGV link{]}}

\textbf{VHGC\_12{ }chr15:65423177{ }TP{ }12{ }9{ }21{ }0.5{ }0.5{
}303.77{ }Heterozygous{ }VeryHighGC{ }SNP{ }{[}IGV link{]}}

\textbf{VHGC\_11{ }chr9:109497831{ }TP{ }23{ }15{ }38{ }0.5{ }0.5{
}309.77{ }Heterozygous{ }VeryHighGC{ }SNP{ }{[}IGV link{]}}

\textbf{VHGC\_15{ }chr5:180291609{ }TP{ }5{ }17{ }23{ }0.5{ }0.5{
}562.77{ }Heterozygous{ }VeryHighGC{ }SNP{ }{[}IGV link{]}}

\textbf{VLGC\_01{ }chr12:37587951{ }TP{ }18{ }20{ }38{ }0.5{ }0.5{
}723.77{ }Heterozygous{ }VeryLowGC{ }SNP{ }{[}IGV link{]}}

\textbf{VLGC\_10{ }chr18:66964143{ }TP{ }27{ }9{ }36{ }0.5{ }0.5{
}261.77{ }Heterozygous{ }VeryLowGC{ }SNP{ }{[}IGV link{]}}

\textbf{VLGC\_02{ }chr12:87708419{ }TP{ }14{ }31{ }45{ }0.5{ }0.5{
}1125.77{ }Heterozygous{ }VeryLowGC{ }SNP{ }{[}IGV link{]}}

\textbf{VLGC\_09{ }chr2:125541813{ }TP{ }19{ }10{ }29{ }0.5{ }0.5{
}331.77{ }Heterozygous{ }VeryLowGC{ }SNP{ }{[}IGV link{]}}

\textbf{VLGC\_06{ }chr6:152024872{ }TP{ }26{ }30{ }56{ }0.5{ }0.5{
}1112.77{ }Heterozygous{ }VeryLowGC{ }SNP{ }{[}IGV link{]}}

\textbf{}\\

\textbf{CANCER VARIANTS}

CS\_011{ }chr1:43349338{ }0.5{ }{ }COSM18918{~{ } }MPL{~}

CS\_018{ }chr1:64844834{ }0.5{ }{ }COSM41751{~{ } }JAK1{~}

CS\_022{ }chr1:114421610{ }0.25{ }{ }COSM1559266{~{ } }TRIM33{~}

CS\_027{ }chr1:114713908{ }0.015625{ }{ }COSM583{~{ } }NRAS{~}

CS\_040{ }chr1:114716127{ }0.0009766{ }COSM563{~{ } }NRAS{~}

CS\_064{ }chr1:226064434{ }0.25{ }{ }COSM327928{~{ } }H3F3A{~}

CS\_071{ }chr10:43121968{ }0.25{ }{ }COSM965{~{ } }RET{~}

CS\_075{ }chr10:86923490{ }1{ }{ }COSM1349419{~{ } }BMPR1A{~}

CS\_077{ }chr10:87933147{ }0.25{ }{ }COSM5219{~{ } }PTEN{~}

CS\_080{ }chr10:87957915{ }0.0625{ }{ }COSM5154{~{ } }PTEN{~}

CI\_012{ }chr10:87961054{ }0.015625{ }{ }COSM5801{~{ } }PTEN{~}

CS\_088{ }chr10:121520163{ }0.25{ }{ }COSM36903{~{ } }FGFR2{~}

CS\_091{ }chr11:533874{ }0.015625{ }{ }COSM499{~{ } }HRAS{~}

CI\_015{ }chr11:32396361{ }0.015625{ }{ }COSM21392{~{ } }WT1{~}

CS\_127{ }chr11:119278181{ }0.125{ }{ }COSM34052{~{ } }CBL{~}

CI\_019{ }chr12:6667903{ }1{ }{ }COSM431727{~{ } }ZNF384{~}

CS\_178{ }chr12:25245351{ }0.03125{ }{ }COSM516{~{ } }KRAS{~}

CS\_196{ }chr12:57751647{ }0.5{ }{ }COSM363684{~{ } }CDK4{~}

CS\_203{ }chr12:111418845{ }0.125{ }{ }COSM1235365{~{ } }SH2B3{~}

CS\_209{ }chr12:112450395{ }0.125{ }{ }COSM13015{~{ } }PTPN11{~}

CI\_021{ }chr12:120994314{ }0.0039063{ }COSM21481{~{ } }HNF1A{~}

CS\_223{ }chr13:28018505{ }0.125{ }{ }COSM783{~{ } }FLT3{~}

CS\_225{ }chr13:32339399{ }0.03125{ }{ }COSM1366454{~{ } }BRCA2{~}

CS\_229{ }chr13:48380086{ }0.125{ }{ }COSM1559416{~{ } }RB1{~}

CS\_247{ }chr14:104780214{ }1{ }{ }COSM33765{~{ } }AKT1{~}

CS\_261{ }chr15:66481793{ }0.0625{ }{ }COSM232755{~{ } }MAP2K1{~}

CS\_262{ }chr15:66485086{ }0.125{ }{ }COSM224488{~{ } }MAP2K1{~}

CS\_268{ }chr15:90088606{ }0.125{ }{ }COSM33733{~{ } }IDH2{~}

CI\_025{ }chr16:2071904{ }0.0625{ }{ }COSM1583165{~{ } }TSC2{~}

CS\_278{ }chr16:2175555{ }0.25{ }{ }COSM1578119{~{ } }TRAF7{~}

CS\_296{ }chr16:68812184{ }0.0625{ }{ }COSM1379165{~{ } }CDH1{~}

CS\_308{ }chr17:7673803{ }0.0625{ }{ }COSM10659{~{ } }TP53{~}

CS\_326{ }chr17:12107827{ }0.25{ }{ }COSM26105{~{ } }MAP2K4{~}

CS\_336{ }chr17:39724747{ }0.0625{ }{ }COSM14062{~{ } }ERBB2{~}

CS\_341{ }chr17:42207711{ }0.5{ }{ }COSM1716590{~{ } }STAT5B{~}

CS\_342{ }chr17:42322402{ }0.0625{ }{ }COSM1155739{~{ } }STAT3{~}

CS\_344{ }chr17:43104120{ }0.0625{ }{ }COSM1559465{~{ } }BRCA1{~}

CS\_363{ }chr18:44951942{ }0.0039063{ }COSM1318400{~{ } }SETBP1{~}

CS\_371{ }chr18:51065549{ }0.5{ }{ }COSM14122{~{ } }SMAD4{~}

CS\_376{ }chr19:1223126{ }0.5{ }{ }COSM21360{~{ } }STK11{~}

CS\_380{ }chr19:3118944{ }0.0078125{ }COSM52969{~{ } }GNA11{~}

CS\_383{ }chr19:4363794{ }0.25{ }COSM328588{~{ } }SH3GL1{~}

CI\_034{ }chr19:33302346{ }1{ }COSM18922{~{ } }CEBPA{~}

CS\_415{ }chr2:25234374{ }0.0625{ }COSM53042{~{ } }DNMT3A{~}

CS\_421{ }chr2:29213994{ }0.125{ }COSM28499{~{ } }ALK{~}

CS\_429{ }chr2:29223430{ }0.015625{ }COSM28502{~{ } }ALK{~}

CI\_035{ }chr2:47803500{ }0.25{ }COSM13395{~{ } }MSH6{~}

CS\_449{ }chr2:197401887{ }1{ }COSM145923{~{ } }SF3B1{~}

CS\_457{ }chr2:208248389{ }0.0009766{ }COSM28749{~{ } }IDH1{~}

CS\_462{ }chr2:222918051{ }0.0039063{ }COSM1494811{~{ } }ACSL3{~}

CS\_471{ }chr21:33027787{ }0.03125{ }COSM1559585{~{ } }OLIG2{~}

CS\_475{ }chr21:43104346{ }0.03125{ }COSM166866{~{ } }U2AF1{~}

CS\_484{ }chr22:28695793{ }0.0019531{ }COSM35476{~{ } }CHEK2{~}

CS\_488{ }chr22:29636805{ }0.0019531{ }COSM21991{~{ } }NF2{~}

CS\_494{ }chr3:10142050{ }0.5{ }COSM14372{~{ } }VHL{~}

CS\_499{ }chr3:37042331{ }1{ }COSM25900{~{ } }MLH1{~}

CS\_500{ }chr3:38141150{ }0.0078125{ }COSM85940{~{ } }MYD88{~}

CS\_515{ }chr3:41224646{ }0.0078125{ }COSM5667{~{ } }CTNNB1{~}

CS\_526{ }chr3:128481887{ }0.0078125{ }COSM41611{~{ } }GATA2{~}

CS\_542{ }chr3:179218303{ }0.0078125{ }COSM763{~{ } }PIK3CA{~}

CS\_560{ }chr3:179234297{ }0.0019531{ }COSM776{~{ } }PIK3CA{~}

CS\_572{ }chr4:1801841{ }0.03125{ }COSM715{~{ } }FGFR3{~}

CS\_573{ }chr4:1804362{ }0.015625{ }COSM716{~{ } }FGFR3{~}

CS\_585{ }chr4:54014115{ }0.0039063{ }COSM159988{~{ } }CHIC2{~}

CI\_044{ }chr4:54274884{ }0.0009766{ }COSM12418{~{ } }PDGFRA{~}

CS\_594{ }chr4:54727444{ }0.0078125{ }COSM1252{~{ } }KIT{~}

CS\_608{ }chr4:105235590{ }0.0039063{ }COSM41644{~{ } }TET2{~}

CS\_611{ }chr5:1294988{ }0.03125{ }COSM1717366{~{ } }TERT{~}

CI\_050{ }chr5:35874471{ }0.5{ }COSM214586{~{ } }IL7R{~}

CI\_053{ }chr5:112839514{ }0.125{ }COSM18719{~{ } }APC{~}

CS\_641{ }chr6:26031978{ }0.0039063{ }COSM829202{~{ } }HIST1H3B{~}

CS\_648{ }chr6:37171174{ }0.0039063{ }COSM220740{~{ } }PIM1{~}

CS\_653{ }chr6:106088449{ }0.0039063{ }COSM97162{~{ } }PRDM1{~}

CS\_654{ }chr6:108664138{ }1{ }COSM35364{~{ } }FOXO3{~}

CS\_677{ }chr7:55174015{ }0.015625{ }COSM6239{~{ } }EGFR{~}

CI\_071{ }chr7:55181287{ }0.03125{ }COSM26720{~{ } }EGFR{~}

CS\_689{ }chr7:116783374{ }0.0019531{ }COSM700{~{ } }MET{~}

CS\_693{ }chr7:129206557{ }0.0078125{ }COSM216037{~{ } }SMO{~}

CS\_695{ }chr7:129210500{ }0.0019531{ }COSM13146{~{ } }SMO{~}

CS\_703{ }chr7:140753336{ }0.015625{ }COSM476{~{ } }BRAF{~}

CS\_725{ }chr8:38428373{ }0.0078125{ }COSM1456955{~{ } }FGFR1{~}

CI\_077{ }chr9:5070033{ }0.5{ }COSM24440{~{ } }JAK2{~}

CS\_743{ }chr9:5073770{ }0.0019531{ }COSM12600{~{ } }JAK2{~}

CS\_754{ }chr9:21971187{ }0.0009766{ }COSM12473{~{ } }CDKN2A{~}

CS\_756{ }chr9:37015168{ }0.0009766{ }COSM85953{~{ } }PAX5{~}

CS\_759{ }chr9:77794572{ }0.0009766{ }COSM28757{~{ } }GNAQ{~}

CS\_766{ }chr9:107487067{ }1{ }COSM248828{~{ } }KLF4{~}

CS\_771{ }chr9:130872896{ }0.03125{ }COSM12560{~{ } }ABL1{~}

CI\_081{ }chr9:132905669{ }0.0078125{ }COSM1636659{~{ } }TSC1{~}

CI\_084{ }chr9:136496196{ }0.03125{ }COSM12774{~{ } }NOTCH1{~}

CS\_775{ }chr9:136504892{ }0.015625{ }COSM12771{~{ } }NOTCH1{~}

CS\_778{ }chrX:1196852{ }0.0009766{ }COSM41268{~{ } }CRLF2{~}

CI\_085{ }chrX:48791198{ }0.0019531{ }COSM13225{~{ } }GATA1{~}

CS\_790{ }chrX:71119403{ }0.0009766{ }COSM131593{~{ } }MED12{~}

CI\_086{ }chrX:77664683{ }0.0019531{ }COSM250047{~{ } }ATRX

\textbf{MICROSATELLITES\\
}

\textbf{STRUCTURAL VARIANTS}

DEL\_01

DEL\_02

DEL\_03

DEL\_04

DEL\_05

DEL\_07

DEL\_09

DEL\_12

DEL\_13

DEL\_14

DUP\_01

DUP\_02

DUP\_03

DUP\_04

DUP\_05

DUP\_06

DUP\_07

DUP\_08

DUP\_09

DUP\_10

DUP\_11

DUP\_12

INS\_01

INS\_02

INS\_06

INS\_07

INS\_08

INS\_10

INS\_11

INS\_12

INS\_14

INV\_01

INV\_02

INV\_05

INV\_06

INV\_08

INV\_09

INV\_11

INV\_12

INV\_13

INV\_14

MEI\_01

MEI\_02

MEI\_03

MEI\_05

MEI\_06

\end{document}
