---
title: "Anaquin: TransQuin Report"
header-includes: \usepackage{graphicx}
output: 
    pdf_document:
        keep_tex: true
        toc: yes
        toc_depth: 2
---


\pagebreak


# TransQuin Alignment

## Summary statistics for: A1

```{ eval=FALSE}
Summary for file: K_RMXA1v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 11440146 (23.8709%) reads
   Synthetic:  36484961 (76.1291%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      85464855 exons
   Query:      32195352 introns
   Query:      161775 bases

   Dilution:   0.761291

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.997479
   Specificity: 0.976043
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.993191
   Specificity: 0.840372
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.691812
   Specificity: 0.932919
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.002521
   Intron: 0.006809
   Gene:   0.026316



```


\pagebreak


## Sequin statistics for: A1

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.845083	1	0.999532	1	1	0.845083	0.998971
R1_102	0.721599	1	0.994156	1	0.901155	0.721599	0.998544
R1_103	0.628255	1	0.995319	1	0.977373	0.628255	0.99561
R1_11	0.645833	1	0.998799	1	0.998818	0.645833	0.997923
R1_12	0.591978	1	0.992884	1	0.9983	0.591978	0.997669
R1_13	0.920394	1	0.987685	0.909091	0.946898	0.920394	0.995898
R1_14	1	0.999823	--	--	1	0.995502	1
R1_21	0.630945	1	0.993234	1	0.916456	0.630945	0.942446
R1_22	0.528054	1	0.998358	1	0.0416689	0.528054	0.993191
R1_23	0.608499	1	0.988379	1	0.00621383	0.608499	0.997146
R1_24	0.999782	1	0.995249	1	0.998266	0.999782	0.997819
R1_31	0.688293	1	0.995473	1	0.244225	0.688293	0.994565
R1_32	0.545455	1	0.999737	1	1	0.545455	0.99894
R1_33	0.899202	1	0.993455	1	0.994169	0.899202	0.998359
R1_41	0.78125	1	0.996675	1	0.999808	0.78125	0.995575
R1_42	0.617479	1	0.980681	1	0.976687	0.617479	0.988844
R1_43	0.540369	1	0.990501	0.973684	0.96341	0.540369	0.996994
R1_51	0.607103	1	0.996288	1	0.999373	0.607103	0.993542
R1_52	0.622807	1	0.991475	1	1	0.622807	0.971446
R1_53	0.998489	1	0.99784	1	0.370359	0.998489	0.996482
R1_61	0.679466	1	1	1	1	0.679466	1
R1_62	0.774336	1	1	1	1	0.774336	1
R1_63	0.669443	1	0.995492	1	0.971203	0.669443	0.995741
R1_71	0.740968	1	0.994243	1	0.998708	0.740968	0.987705
R1_72	0.582844	1	1	1	1	0.582844	1
R1_73	0.731952	1	0.984041	1	0.98719	0.731952	0.996865
R1_81	0.747244	1	0.99754	1	0.995044	0.747244	0.996849
R1_82	0.587741	1	0.995744	1	0.992876	0.587741	0.991501
R1_83	0.644813	1	0.998373	1	0.997866	0.644813	0.997613
R1_91	0.685055	1	1	1	1	0.685055	1
R1_92	0.777919	1	0.980772	1	0.995753	0.777919	0.99513
R1_93	0.625086	1	0.996054	1	0.999554	0.625086	0.996156
R2_1	1	1	--	--	0.99196	1	0.99196
R2_105	1	0.967742	--	--	0.874667	0.99696	0.874667
R2_115	0.837412	1	0.988931	1	0.996079	0.837412	0.923815
R2_116	0.584726	1	0.987644	1	1	0.584726	0.997557
R2_117	0.711066	1	0.999248	1	0.998209	0.711066	0.996411
R2_14	0.719569	1	0.988252	1	0.987247	0.719569	0.991507
R2_150	0.835063	1	0.998167	1	0.518953	0.835063	0.991781
R2_151	0.578303	1	0.983911	1	0.000681453	0.578303	0.999244
R2_152	0.603856	1	0.99729	1	0.0120943	0.603856	0.998386
R2_153	0.677873	1	0.995311	1	0.359343	0.677873	0.999099
R2_154	0.855658	1	0.989621	1	0.547718	0.855658	0.994631
R2_18	0.628326	1	0.9853	1	0.852844	0.628326	0.985276
R2_19	0.811493	1	0.992328	1	0.98788	0.811493	0.998432
R2_20	0.735871	1	0.997986	1	0.997017	0.735871	0.997394
R2_24	0.586003	0.979592	0.991056	0.957447	0.693273	0.586003	0.998239
R2_26	0.916638	1	0.994978	1	0.994222	0.916638	0.997942
R2_27	0.754167	1	0.995777	1	1	0.754167	0.999606
R2_28	0.68506	1	1	1	1	0.68506	1
R2_32	0.671486	1	0.98234	1	1	0.671486	0.997715
R2_33	0.879859	1	1	1	0.000191773	0.879859	1
R2_37	0.719465	1	0.993295	1	0.982877	0.719465	0.998428
R2_38	0.391233	1	0.989899	1	1	0.391233	0.995805
R2_41	0.837384	1	0.991692	1	0.995112	0.837384	0.989443
R2_42	0.882171	1	0.998311	1	1	0.882171	0.999561
R2_45	0.462072	1	1	1	1	0.462072	1
R2_46	0.688515	1	0.993056	1	1	0.688515	0.967061
R2_47	0.887733	1	0.996614	1	0.997939	0.887733	0.998675
R2_53	0.396308	1	0.989455	1	0.997009	0.396308	0.997744
R2_54	0.864703	1	0.982539	1	0.996987	0.864703	0.996936
R2_55	0.880804	1	0.987179	1	0.873955	0.880804	0.99142
R2_57	0.65303	1	0.996765	1	1	0.65303	0.998819
R2_59	0.588824	1	0.988263	1	0.993952	0.588824	0.999726
R2_6	0.702025	1	0.996212	1	0.990037	0.702025	0.98913
R2_60	0.605403	1	0.996636	1	1	0.605403	0.997073
R2_63	0.879982	1	0.987448	1	0.997983	0.879982	0.997517
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.529254	1	0.999748	1	0.999936	0.529254	0.985015
R2_67	0.959847	1	1	1	1	0.959847	1
R2_68	0.705317	1	0.998856	1	1	0.705317	0.999367
R2_7	0.760474	1	0.994878	1	0.994215	0.760474	0.994466
R2_71	0.72861	1	0.986272	1	1	0.72861	0.998168
R2_72	0.320416	0.866667	0.993094	0.727273	1	0.320416	0.99815
R2_73	0.94562	1	0.988466	1	0.996508	0.94562	0.999279
R2_76	0.470652	1	1	1	1	0.470652	1

```


\pagebreak


## Summary statistics for: A2

```{ eval=FALSE}
Summary for file: K_RMXA2v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 10700630 (23.3808%) reads
   Synthetic:  35066089 (76.6192%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      82814262 exons
   Query:      34992686 introns
   Query:      162619 bases

   Dilution:   0.766192

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.997479
   Specificity: 0.974098
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.986381
   Specificity: 0.745193
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.691909
   Specificity: 0.928206
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.002521
   Intron: 0.013619
   Gene:   0.026316



```


\pagebreak


## Sequin statistics for: A2

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.843342	1	0.997495	1	0.995816	0.843342	0.995889
R1_102	0.721599	1	0.995219	1	0.00552418	0.721599	0.998544
R1_103	0.628532	1	0.995586	1	0.185487	0.628532	0.995612
R1_11	0.647177	1	0.999245	1	0.0529042	0.647177	0.997927
R1_12	0.590595	1	0.9933	0.956522	1	0.590595	0.998247
R1_13	0.920394	1	0.988579	0.909091	0.661924	0.920394	0.995082
R1_14	1	0.999844	--	--	1	0.991045	1
R1_21	0.630945	1	0.994394	1	0.942939	0.630945	0.942446
R1_22	0.527602	1	0.998493	1	0.101617	0.527602	0.994032
R1_23	0.604667	1	0.989406	1	0.0153541	0.604667	0.997701
R1_24	1	1	0.995351	1	0.995767	1	0.99782
R1_31	0.687823	1	0.994739	1	0.998694	0.687823	0.995916
R1_32	0.545455	1	0.998386	1	1	0.545455	0.995772
R1_33	0.911321	1	0.998728	1	0.99726	0.911321	0.999676
R1_41	0.78125	1	0.996939	1	0.932387	0.78125	0.994695
R1_42	0.617479	1	0.981914	1	0.891942	0.617479	0.986842
R1_43	0.541274	1	0.990613	0.973684	0.989699	0.541274	0.996335
R1_51	0.607103	1	0.995118	1	0.988762	0.607103	0.99446
R1_52	0.623188	1	0.979032	1	0.992218	0.623188	0.977858
R1_53	0.998489	1	0.995792	1	0.726932	0.998489	0.996983
R1_61	0.676962	1	0.99811	1	0.00262261	0.676962	0.998768
R1_62	0.767383	1	0.992147	1	0.00112939	0.767383	0.999177
R1_63	0.669964	1	0.995761	1	0.964477	0.669964	0.995745
R1_71	0.740968	1	0.993449	1	0.964899	0.740968	0.987705
R1_72	0.600451	1	1	1	0.000569401	0.600451	1
R1_73	0.732371	1	0.989559	0.96	0.531614	0.732371	0.996583
R1_81	0.747244	1	0.998029	1	0.994022	0.747244	0.996849
R1_82	0.587741	1	0.992549	1	0.995822	0.587741	0.991501
R1_83	0.645199	1	0.999418	1	0.999628	0.645199	0.99702
R1_91	0.685696	1	1	1	1	0.685696	1
R1_92	0.781726	1	0.980355	1	0.995561	0.781726	0.995153
R1_93	0.625086	1	0.996684	1	0.998378	0.625086	0.996156
R2_1	1	1	--	--	0.98995	1	0.98995
R2_105	1	1	--	--	0.68	1	0.68
R2_115	0.840131	1	0.989271	1	0.996256	0.840131	0.99549
R2_116	0.524582	1	0.996665	1	1	0.524582	0.998183
R2_117	0.710041	1	0.999708	1	0.999328	0.710041	0.997122
R2_14	0.719569	1	0.988161	1	0.737964	0.719569	0.991507
R2_150	0.835448	1	0.998619	1	0.673702	0.835448	0.991785
R2_151	0.565179	1	0.981308	0.85	0.00120631	0.565179	0.998454
R2_152	0.603368	1	0.998585	1	0.028801	0.603368	0.997981
R2_153	0.669927	1	0.981132	1	1	0.669927	0.999088
R2_154	0.855658	1	0.986231	1	0.753071	0.855658	0.994631
R2_18	0.628326	1	0.982418	1	0.932695	0.628326	0.986486
R2_19	0.811493	1	0.994275	1	0.395306	0.811493	0.998717
R2_20	0.732026	1	0.998159	1	0.998462	0.732026	0.998951
R2_24	0.586299	0.979592	0.987268	0.957447	0.764632	0.586299	0.998994
R2_26	0.916638	1	0.994996	1	0.995074	0.916638	0.997755
R2_27	0.750595	1	0.994545	1	1	0.750595	0.999208
R2_28	0.686747	1	0.996957	1	1	0.686747	0.979381
R2_32	0.671486	1	0.992126	1	1	0.671486	0.998171
R2_33	0.90106	1	1	1	1	0.90106	1
R2_37	0.719239	1	1	1	1	0.719239	1
R2_38	0.399143	1	0.984831	1	1	0.399143	0.998351
R2_41	0.839171	1	0.991918	1	0.998755	0.839171	0.990299
R2_42	0.874031	1	1	1	1	0.874031	1
R2_45	0.464239	1	0.999817	1	0.998319	0.464239	0.999689
R2_46	0.70475	1	0.988748	1	1	0.70475	0.967396
R2_47	0.887733	1	0.997391	1	0.998014	0.887733	0.998675
R2_53	0.396487	1	0.997205	1	0.997577	0.396487	0.999097
R2_54	0.864703	1	0.983372	1	0.996267	0.864703	0.997191
R2_55	0.880804	1	0.987741	1	0.862818	0.880804	0.990647
R2_57	0.659591	1	0.996748	1	0.00525698	0.659591	0.998248
R2_59	0.583656	1	1	1	1	0.583656	1
R2_6	0.702989	1	0.998334	1	0.994613	0.702989	0.995902
R2_60	0.604692	1	0.996744	1	0.00539659	0.604692	0.997069
R2_63	0.884363	1	0.984499	1	0.985809	0.884363	0.997037
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.529254	1	0.999748	1	0.999958	0.529254	0.985015
R2_67	0.992352	1	1	1	1	0.992352	1
R2_68	0.707775	1	0.996954	1	1	0.707775	0.999684
R2_7	0.760474	1	0.997273	1	0.999218	0.760474	0.993367
R2_71	0.69385	1	0.992081	1	0.992248	0.69385	0.997118
R2_72	0.338827	0.866667	0.985673	0.545455	1	0.338827	0.999124
R2_73	0.950398	1	0.991652	1	0.99652	0.950398	0.975023
R2_76	0.477536	1	0.99809	1	1	0.477536	0.999242

```


\pagebreak


## Summary statistics for: A3

```{ eval=FALSE}
Summary for file: K_RMXA3v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 9781841 (22.7411%) reads
   Synthetic:  33232155 (77.2589%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      78415783 exons
   Query:      29409848 introns
   Query:      166617 bases

   Dilution:   0.772589

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.997479
   Specificity: 0.971556
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.992218
   Specificity: 0.840566
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.696877
   Specificity: 0.912440
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.002521
   Intron: 0.007782
   Gene:   0.026316



```


\pagebreak


## Sequin statistics for: A3

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.844212	1	0.999087	1	1	0.844212	0.99897
R1_102	0.723352	1	0.995112	1	0.0284201	0.723352	0.998065
R1_103	0.629086	1	0.995227	1	0.535702	0.629086	0.99518
R1_11	0.646505	1	0.998816	1	0.0505829	0.646505	0.997925
R1_12	0.591978	1	0.993087	1	0.999141	0.591978	0.997669
R1_13	0.920394	1	0.988092	0.909091	0.909266	0.920394	0.995082
R1_14	1	0.999779	--	--	1	0.995502	1
R1_21	0.630945	1	0.993913	1	0.951111	0.630945	0.941599
R1_22	0.528054	1	0.998712	1	0.201023	0.528054	0.994037
R1_23	0.610589	1	0.987265	1	0.0338936	0.610589	0.997156
R1_24	1	1	0.995487	1	0.998898	1	0.99782
R1_31	0.687823	1	0.994534	1	0.998714	0.687823	0.995916
R1_32	0.544876	1	0.997818	1	1	0.544876	0.995767
R1_33	0.912504	1	0.997554	1	0.996416	0.912504	0.998706
R1_41	0.78125	1	0.99667	1	0.999697	0.78125	0.996457
R1_42	0.617479	1	0.980701	1	0.897666	0.617479	0.985844
R1_43	0.540731	1	0.99053	0.973684	0.984815	0.540731	0.996331
R1_51	0.607103	1	0.996042	1	0.992969	0.607103	0.993542
R1_52	0.623951	1	0.989971	1	1	0.623951	0.969769
R1_53	0.998489	1	0.998404	1	0.973515	0.998489	0.996983
R1_61	0.709516	1	0.993926	1	1	0.709516	0.998825
R1_62	0.776233	1	1	1	1	0.776233	1
R1_63	0.669703	1	0.995554	1	0.974728	0.669703	0.995743
R1_71	0.740968	1	0.993971	1	0.988434	0.740968	0.985685
R1_72	0.604966	1	1	1	1	0.604966	1
R1_73	0.731743	1	0.983855	0.96	0.986432	0.731743	0.996864
R1_81	0.747638	1	0.996786	1	0.994027	0.747638	0.99685
R1_82	0.587741	1	0.992415	1	0.997628	0.587741	0.993612
R1_83	0.644427	1	0.997372	1	0.996722	0.644427	0.998208
R1_91	0.684413	1	0.998495	1	1	0.684413	0.998129
R1_92	0.778765	1	0.980597	1	0.995994	0.778765	0.995135
R1_93	0.625086	1	0.995437	1	0.998834	0.625086	0.996703
R2_1	1	1	--	--	0.994975	1	0.994975
R2_105	1	0.979167	--	--	0.981333	0.994595	0.981333
R2_115	0.837955	1	0.989482	1	0.997475	0.837955	0.923861
R2_116	0.583771	1	0.988752	1	1	0.583771	0.997553
R2_117	0.709016	1	0.999618	1	0.99934	0.709016	0.9964
R2_14	0.719569	1	0.989328	1	0.838494	0.719569	0.991507
R2_150	0.835832	1	0.997978	1	0.821556	0.835832	0.996334
R2_151	0.589676	1	0.990974	1	0.00596157	0.589676	0.997779
R2_152	0.603368	1	0.998327	1	0.0588556	0.603368	0.997981
R2_153	0.685208	1	0.995346	1	0.983165	0.685208	0.998219
R2_154	0.855658	1	0.986493	1	0.834231	0.855658	0.994631
R2_18	0.628326	1	0.984271	1	0.935806	0.628326	0.986486
R2_19	0.811493	1	0.994731	1	0.753652	0.811493	0.998574
R2_20	0.733948	1	0.996812	1	0.99733	0.733948	0.996347
R2_24	0.586741	0.979592	0.990687	0.957447	0.842066	0.586741	0.998743
R2_26	0.916982	1	0.995763	1	0.994951	0.916982	0.997942
R2_27	0.755357	1	0.993405	1	0.999065	0.755357	0.998426
R2_28	0.68988	1	0.999127	1	1	0.68988	0.980144
R2_32	0.671486	1	0.989034	1	1	0.671486	0.988678
R2_33	0.971731	1	0.947368	1	1	0.971731	0.99278
R2_37	0.727623	1	0.993796	1	0.993171	0.727623	0.997825
R2_38	0.399473	1	0.995508	1	1	0.399473	0.998353
R2_41	0.839171	1	0.992037	1	0.998985	0.839171	0.968247
R2_42	0.885659	1	0.997419	1	1	0.885659	0.949709
R2_45	0.464528	1	0.997782	1	1	0.464528	0.998447
R2_46	0.719784	1	0.99254	1	1	0.719784	0.967273
R2_47	0.888518	1	0.996762	1	0.996155	0.888518	0.998676
R2_53	0.397025	1	0.994321	1	0.994027	0.397025	0.997748
R2_54	0.864703	1	0.982499	1	0.997465	0.864703	0.997446
R2_55	0.880804	1	0.988265	1	0.838564	0.880804	0.990647
R2_57	0.659205	1	0.99747	1	0.00678764	0.659205	0.99883
R2_59	0.596415	1	0.9968	1	0.998424	0.596415	0.999459
R2_6	0.702025	1	0.995403	1	0.991333	0.702025	0.993179
R2_60	0.605759	1	0.995309	1	0.00940813	0.605759	0.997074
R2_63	0.884801	1	0.985674	1	0.988104	0.884801	0.996547
R2_65	1	0.997792	--	--	0.994493	0.998894	0.994493
R2_66	0.529254	1	0.988763	1	0.999892	0.529254	0.986
R2_67	0.967495	1	1	1	1	0.967495	1
R2_68	0.717828	1	0.996747	1	0.998457	0.717828	0.999378
R2_7	0.76132	1	0.99719	1	0.999558	0.76132	0.994472
R2_71	0.916444	1	0.993642	1	0.997036	0.916444	0.986331
R2_72	0.347439	0.866667	0.991036	0.727273	1	0.347439	0.997017
R2_73	0.943572	1	0.985888	1	0.998646	0.943572	0.999759
R2_76	0.477536	1	0.994867	1	1	0.477536	0.998485

```


\pagebreak


## Summary statistics for: B1

```{ eval=FALSE}
Summary for file: G_RMXB1v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 6299276 (15.7506%) reads
   Synthetic:  33694649 (84.2494%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      86364151 exons
   Query:      42105718 introns
   Query:      158128 bases

   Dilution:   0.842494

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.993277
   Specificity: 0.977562
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.984436
   Specificity: 0.793287
   Detection:   1.88828 (R1_72)

   -------------------- Base level --------------------

   Sensitivity: 0.683749
   Specificity: 0.943312
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.006723
   Intron: 0.015564
   Gene:   0.052632



```


\pagebreak


## Sequin statistics for: B1

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.787641	1	0.996774	1	1	0.787641	0.998896
R1_102	0.721248	1	0.98946	1	0.108223	0.721248	0.999514
R1_103	0.627978	1	0.992978	1	0.959322	0.627978	0.996484
R1_11	0.644489	1	0.989534	1	1	0.644489	0.997919
R1_12	0.569848	1	0.992637	1	0.991467	0.569848	0.998788
R1_13	0.920394	1	0.988299	0.909091	0.983353	0.920394	0.995082
R1_14	1	0.999917	--	--	0.996988	0.996988	0.996988
R1_21	0.630945	1	0.993465	1	0.902076	0.630945	0.941599
R1_22	0.526697	1	0.999028	1	0.0033303	0.526697	0.997429
R1_23	0.609892	1	0.983464	1	0.00570884	0.609892	0.997153
R1_24	0.999345	1	0.995356	1	0.998841	0.999345	0.998472
R1_31	0.688764	1	0.99394	1	0.997605	0.688764	0.992547
R1_32	0.541401	1	1	1	1	0.541401	1
R1_33	0.88324	1	0.997831	1	1	0.88324	0.999331
R1_41	0.78125	1	0.99979	1	0.997007	0.78125	0.995575
R1_42	0.617479	1	0.989977	1	0.975927	0.617479	0.990854
R1_43	0.541999	1	0.986484	0.973684	0.348855	0.541999	0.996339
R1_51	0.603157	1	0.995454	1	0.855858	0.603157	0.995349
R1_52	0.622807	1	0.994252	1	1	0.622807	0.998166
R1_53	0.998993	1	0.997143	1	0.718094	0.998993	0.996985
R1_61	0.718698	1	1	1	1	0.718698	1
R1_62	0.697219	1	1	1	1	0.697219	1
R1_63	0.670484	1	0.994539	1	0.857561	0.670484	0.995748
R1_71	0.740968	1	0.99358	1	0.848788	0.740968	0.986694
R1_72	0.544921	1	1	1	1	0.544921	1
R1_73	0.731115	1	0.991175	0.96	0.993206	0.731115	0.997431
R1_81	0.747244	1	0.996891	1	0.991853	0.747244	0.996325
R1_82	0.587741	1	0.999352	1	0.987979	0.587741	0.992908
R1_83	0.647513	1	0.998826	1	0.998681	0.647513	0.998216
R1_91	0.66517	1	1	1	1	0.66517	1
R1_92	0.778765	1	0.980861	1	0.995714	0.778765	0.995135
R1_93	0.625086	1	0.996681	1	0.99831	0.625086	0.996156
R2_1	1	1	--	--	0.994975	1	0.994975
R2_105	1	0.992857	--	--	0.946667	0.994398	0.946667
R2_115	0.842849	1	0.991243	1	0.99245	0.842849	0.923719
R2_116	0.573747	1	0.984551	1	1	0.573747	0.999169
R2_117	0.71209	1	0.998192	1	0.997355	0.71209	0.994989
R2_14	0.719569	1	0.984461	1	0.904376	0.719569	0.991507
R2_150	0.831603	1	0.999201	1	0.0928247	0.831603	0.997694
R2_151	0.56343	1	1	1	0.000648549	0.56343	1
R2_152	0.606541	1	0.99787	1	0.0104632	0.606541	0.997992
R2_153	0.680318	1	0.991786	1	1	0.680318	0.988016
R2_154	0.853349	1	0.98975	1	0.345297	0.853349	0.994616
R2_18	0.628326	1	0.98623	1	0.852354	0.628326	0.986486
R2_19	0.811146	1	0.99566	1	0.940862	0.811146	0.999001
R2_20	0.733948	1	0.998092	1	0.995501	0.733948	0.997388
R2_24	0.586003	0.979592	0.990261	0.957447	0.846247	0.586003	0.998239
R2_26	0.916982	1	0.992369	1	0.957098	0.916982	0.997569
R2_27	0.752976	1	0.992291	1	1	0.752976	0.99921
R2_28	0.689157	1	1	1	1	0.689157	1
R2_32	0.66964	1	0.996432	1	1	0.66964	0.998624
R2_33	0.416961	0.5	1	0	nan	0.416961	1
R2_37	0.694992	1	0.994852	1	0.981481	0.694992	0.999348
R2_38	0.379038	1	0.992625	1	1	0.379038	0.997398
R2_41	0.837026	1	0.993443	1	0.9981	0.837026	0.990694
R2_42	0.874031	1	1	1	1	0.874031	1
R2_45	0.462216	1	0.996477	1	1	0.462216	0.998439
R2_46	0.702646	1	1	1	1	0.702646	1
R2_47	0.888911	1	0.996626	1	0.995795	0.888911	0.998237
R2_53	0.387345	1	0.995233	1	1	0.387345	0.999075
R2_54	0.864703	1	0.985219	1	0.996922	0.864703	0.997701
R2_55	0.880804	1	0.983178	1	0.99546	0.880804	0.990647
R2_57	0.625241	1	1	0.941176	1	0.625241	1
R2_59	0.553456	1	0.993644	1	0.984615	0.553456	0.999708
R2_6	0.702989	1	0.996895	1	0.999525	0.702989	0.990489
R2_60	0.607892	1	0.997855	1	0.998146	0.607892	0.997085
R2_63	0.886991	1	0.998694	1	0.999702	0.886991	0.997046
R2_65	1	1	--	--	0.993392	1	0.993392
R2_66	0.528717	1	0.99979	1	0.999945	0.528717	0.987964
R2_67	0.782027	1	1	1	1	0.782027	1
R2_68	0.568811	0.846154	0.996553	0.833333	0.998019	0.568811	0.998823
R2_7	0.76132	1	0.996266	1	0.999624	0.76132	0.994472
R2_71	0.77139	1	0.977113	1	0.975684	0.77139	0.99827
R2_72	0.313586	0.866667	0.985308	0.636364	1	0.313586	0.99811
R2_73	0.921274	1	0.989642	0.952381	0.997619	0.921274	0.999506
R2_76	0.463043	1	0.996114	1	1	0.463043	0.999218

```


\pagebreak


## Summary statistics for: B2

```{ eval=FALSE}
Summary for file: G_RMXB2v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 6434036 (16.0152%) reads
   Synthetic:  33740540 (83.9848%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      85746594 exons
   Query:      43266445 introns
   Query:      157833 bases

   Dilution:   0.839848

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.994118
   Specificity: 0.976567
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.986381
   Specificity: 0.756517
   Detection:   1.88828 (R1_72)

   -------------------- Base level --------------------

   Sensitivity: 0.685982
   Specificity: 0.948160
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.005882
   Intron: 0.013619
   Gene:   0.039474



```


\pagebreak


## Sequin statistics for: B2

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.832898	1	1	1	1	0.832898	1
R1_102	0.698107	1	0.985276	1	0.244275	0.698107	0.999498
R1_103	0.628532	1	0.993265	1	0.970835	0.628532	0.996924
R1_11	0.644489	1	1	1	1	0.644489	1
R1_12	0.590941	1	0.995051	1	1	0.590941	0.998831
R1_13	0.920394	1	0.987888	0.909091	0.995075	0.920394	0.995082
R1_14	1	0.999876	--	--	1	0.995502	1
R1_21	0.630945	1	0.99394	1	0.839282	0.630945	0.942446
R1_22	0.527149	1	0.999506	1	0.0015311	0.527149	0.999142
R1_23	0.609892	1	0.986752	1	0.00298525	0.609892	0.997721
R1_24	0.999563	1	0.995545	1	0.998693	0.999563	0.998037
R1_31	0.688764	1	0.992008	1	0.997879	0.688764	0.99322
R1_32	0.544296	1	1	1	1	0.544296	1
R1_33	0.896837	1	0.983193	1	1	0.896837	0.999341
R1_41	0.78125	1	0.999876	1	0.874219	0.78125	0.995575
R1_42	0.617479	1	0.990803	1	0.985162	0.617479	0.993884
R1_43	0.541636	1	0.985438	0.973684	0.803818	0.541636	0.996005
R1_51	0.604284	1	0.996225	1	0.0829933	0.604284	0.997209
R1_52	0.622807	1	0.994399	1	1	0.622807	0.997557
R1_53	0.998489	1	0.997344	1	0.0360915	0.998489	0.996482
R1_61	0.687813	1	0.986755	1	0.959677	0.687813	0.998788
R1_62	0.686473	1	1	0.8	0.8	0.686473	1
R1_63	0.670484	1	0.993684	1	0.877702	0.670484	0.996133
R1_71	0.740968	1	0.993425	1	0.997341	0.740968	0.984678
R1_72	0.600903	1	1	1	0.8	0.600903	1
R1_73	0.731115	1	0.991099	0.96	0.985267	0.731115	0.997716
R1_81	0.748031	1	0.997639	1	0.992314	0.748031	0.996329
R1_82	0.587741	1	0.99715	1	0.988495	0.587741	0.9908
R1_83	0.647127	1	0.998647	1	0.99802	0.647127	0.998215
R1_91	0.668377	1	1	1	1	0.668377	1
R1_92	0.778342	1	0.983037	1	0.996622	0.778342	0.994595
R1_93	0.625086	1	0.996003	1	0.996222	0.625086	0.996156
R2_1	1	1	--	--	0.99397	1	0.99397
R2_105	1	0.954887	--	--	0.981333	0.994595	0.981333
R2_115	0.837412	1	0.99107	1	0.992674	0.837412	0.923261
R2_116	0.582339	1	0.99	1	1	0.582339	0.997547
R2_117	0.711066	1	0.99903	1	0.997962	0.711066	0.996411
R2_14	0.719569	1	0.984606	1	0.928543	0.719569	0.991507
R2_150	0.82699	1	0.998974	1	0.046194	0.82699	0.993075
R2_151	0.549431	1	0.996042	1	0.000351715	0.549431	0.996035
R2_152	0.603368	1	0.998164	1	0.00536784	0.603368	0.999192
R2_153	0.672983	1	1	1	1	0.672983	1
R2_154	0.855658	1	0.988501	1	0.339719	0.855658	0.994631
R2_18	0.628326	1	0.984501	1	0.951897	0.628326	0.986486
R2_19	0.811146	1	0.996182	1	0.957324	0.811146	0.999001
R2_20	0.733564	1	0.997507	1	0.994503	0.733564	0.997386
R2_24	0.585856	0.979592	0.989461	0.957447	0.828288	0.585856	0.998239
R2_26	0.916982	1	0.99228	1	0.640148	0.916982	0.997569
R2_27	0.748512	1	0.981896	1	1	0.748512	0.999205
R2_28	0.687711	1	1	1	1	0.687711	1
R2_32	0.669948	1	0.993046	1	1	0.669948	0.998624
R2_33	0.95053	1	1	1	1	0.95053	1
R2_37	0.706322	1	0.98032	1	0.981481	0.706322	0.998718
R2_38	0.380026	1	0.97554	1	1	0.380026	0.998268
R2_41	0.836669	1	0.991694	1	0.998627	0.836669	0.995746
R2_42	0.851163	1	0.997821	1	1	0.851163	0.999545
R2_45	0.461061	1	0.996591	1	0.995166	0.461061	0.999061
R2_46	0.694227	1	1	1	1	0.694227	1
R2_47	0.888518	1	0.996794	1	0.975142	0.888518	0.998456
R2_53	0.39111	1	0.999066	1	1	0.39111	0.999542
R2_54	0.864703	1	0.986481	1	0.997388	0.864703	0.997956
R2_55	0.880804	1	0.987022	1	0.998648	0.880804	0.99142
R2_57	0.624855	1	0.993583	0.941176	1	0.624855	0.998766
R2_59	0.559432	1	1	1	1	0.559432	1
R2_6	0.702025	1	0.995921	1	0.998458	0.702025	0.990476
R2_60	0.607892	1	0.997107	1	0.996965	0.607892	0.995923
R2_63	0.886991	1	0.998679	1	0.999706	0.886991	0.997046
R2_65	1	0.984615	--	--	0.994493	0.99779	0.994493
R2_66	0.528717	1	0.999833	1	1	0.528717	0.987964
R2_67	0.944551	1	1	1	1	0.944551	1
R2_68	0.568811	0.846154	0.995703	0.833333	0.996352	0.568811	0.998431
R2_7	0.760474	1	0.994988	1	0.999414	0.760474	0.993367
R2_71	0.676471	1	0.99635	1	1	0.676471	0.999013
R2_72	0.337491	0.866667	0.989529	0.727273	1	0.337491	0.998682
R2_73	0.944937	1	0.986542	1	0.998453	0.944937	0.999519
R2_76	0.470652	1	1	1	1	0.470652	1

```


\pagebreak


## Summary statistics for: B3

```{ eval=FALSE}
Summary for file: G_RMXB3v2.accepted_hits.bam

   Unmapped:   0 reads
   Experiment: 7941201 (17.1508%) reads
   Synthetic:  38361013 (82.8492%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      98520610 exons
   Query:      50973179 introns
   Query:      157888 bases

   Dilution:   0.828492

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.994958
   Specificity: 0.978198
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.983463
   Specificity: 0.751422
   Detection:   1.88828 (R1_72)

   -------------------- Base level --------------------

   Sensitivity: 0.684753
   Specificity: 0.946133
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.005042
   Intron: 0.016537
   Gene:   0.052632



```


\pagebreak


## Sequin statistics for: B3

```{ eval=FALSE}
ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.841601	1	0.99403	1	1	0.841601	0.998967
R1_102	0.717742	1	0.980746	1	0.0221654	0.717742	0.999024
R1_103	0.627978	1	0.992148	1	0.835009	0.627978	0.996922
R1_11	0.645161	1	0.999461	1	1	0.645161	0.998959
R1_12	0.590595	1	0.991325	1	0.99781	0.590595	0.99883
R1_13	0.920394	1	0.988254	0.909091	0.987014	0.920394	0.995898
R1_14	1	0.999656	--	--	1	0.989568	1
R1_21	0.630945	1	0.993922	1	0.822178	0.630945	0.942446
R1_22	0.525792	1	1	1	0.00123868	0.525792	1
R1_23	0.60815	1	0.98571	1	0.00229028	0.60815	0.997714
R1_24	0.999563	1	0.995155	1	0.998935	0.999563	0.997819
R1_31	0.689234	1	0.991529	1	0.983869	0.689234	0.991881
R1_32	0.544296	1	1	1	1	0.544296	1
R1_33	0.890038	1	0.996205	1	1	0.890038	0.999005
R1_41	0.78125	1	0.999946	1	0.996632	0.78125	0.998225
R1_42	0.617479	1	0.991638	1	0.949859	0.617479	0.992872
R1_43	0.541818	1	0.986158	0.973684	0.604672	0.541818	0.996007
R1_51	0.60372	1	0.995606	1	0.831258	0.60372	0.996279
R1_52	0.62357	1	0.993546	1	0.999107	0.62357	0.997559
R1_53	0.998489	1	0.99506	1	0.72554	0.998489	0.995982
R1_61	0.686144	1	1	1	0.911765	0.686144	1
R1_62	0.716182	1	1	1	0.842105	0.716182	1
R1_63	0.670484	1	0.9934	1	0.879043	0.670484	0.995363
R1_71	0.740968	1	0.991256	1	0.984353	0.740968	0.987705
R1_72	0.586456	1	1	1	0.666667	0.586456	1
R1_73	0.731325	1	0.99045	0.96	0.991894	0.731325	0.997432
R1_81	0.749606	1	0.997294	1	0.989216	0.749606	0.996337
R1_82	0.587741	1	0.995803	1	0.987377	0.587741	0.991501
R1_83	0.644813	1	0.998136	1	0.995882	0.644813	0.998209
R1_91	0.679282	1	1	1	1	0.679282	1
R1_92	0.77665	1	0.97995	1	0.996986	0.77665	0.995122
R1_93	0.625086	1	0.996497	1	0.998423	0.625086	0.996156
R2_1	1	1	--	--	0.988945	1	0.988945
R2_105	1	0.983471	--	--	0.984	0.994609	0.984
R2_115	0.842849	1	0.989942	1	0.993289	0.842849	0.92427
R2_116	0.521718	1	0.996266	1	1	0.521718	0.998174
R2_117	0.711066	1	0.999352	1	0.997792	0.711066	0.997126
R2_14	0.719569	1	0.984373	1	0.944812	0.719569	0.991507
R2_150	0.833141	1	0.999017	1	0.0395501	0.833141	0.997698
R2_151	0.584864	1	0.997917	1	0.000290904	0.584864	0.999253
R2_152	0.603856	1	0.997058	1	0.00392735	0.603856	0.998386
R2_153	0.668093	1	0.997268	0.95	1	0.668093	0.99863
R2_154	0.855658	1	0.991371	1	0.158151	0.855658	0.994631
R2_18	0.628326	1	0.985099	1	0.89836	0.628326	0.986486
R2_19	0.811146	1	0.996027	1	0.769173	0.811146	0.998716
R2_20	0.732795	1	0.998371	1	0.998409	0.732795	0.997906
R2_24	0.586151	0.979592	0.988855	0.957447	0.904814	0.586151	0.99824
R2_26	0.916982	1	0.992311	1	0.984013	0.916982	0.997569
R2_27	0.749702	1	0.994169	1	1	0.749702	0.99881
R2_28	0.686747	1	1	1	1	0.686747	1
R2_32	0.667795	1	0.990257	1	1	0.667795	0.997244
R2_33	0.40636	0.5	1	0	0	0.40636	1
R2_37	0.711761	1	0.994798	1	1	0.711761	0.999682
R2_38	0.387607	1	0.9946	1	1	0.387607	0.997455
R2_41	0.837384	1	0.991778	1	0.985197	0.837384	0.989861
R2_42	0.85814	1	1	1	1	0.85814	1
R2_45	0.454703	1	0.999465	1	0.98533	0.454703	0.999682
R2_46	0.683103	1	1	1	1	0.683103	1
R2_47	0.888911	1	0.996685	1	0.995092	0.888911	0.998237
R2_53	0.385374	1	0.998685	1	1	0.385374	0.999535
R2_54	0.864703	1	0.982419	1	0.996371	0.864703	0.997446
R2_55	0.880804	1	0.989737	1	0.973753	0.880804	0.989875
R2_57	0.639907	1	0.99214	0.941176	1	0.639907	0.997593
R2_59	0.562177	1	1	1	1	0.562177	1
R2_6	0.702025	1	0.9962	1	1	0.702025	0.991826
R2_60	0.607892	1	0.996822	1	0.998606	0.607892	0.996503
R2_63	0.886991	1	0.998578	1	1	0.886991	0.996555
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.527107	1	0.999792	1	0.999952	0.527107	0.986935
R2_67	0.804971	1	1	1	1	0.804971	1
R2_68	0.597632	0.923077	0.995682	0.875	0.997295	0.597632	0.99888
R2_7	0.76132	1	0.997063	1	0.99982	0.76132	0.993374
R2_71	0.679813	1	0.998536	1	1	0.679813	0.999018
R2_72	0.300965	0.866667	0.993569	0.454545	1	0.300965	0.999014
R2_73	0.950398	1	0.991451	1	0.997839	0.950398	0.999521
R2_76	0.477536	1	0.990426	1	1	0.477536	0.999242

```


\pagebreak


\pagebreak


# TransQuin Assembly

## Summary statistics for: A1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A1/NotGuided/transcripts.gtf

   Experiment: 156131 features
   Synthetic:  1955 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


## Summary statistics for: A2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A2/NotGuided/transcripts.gtf

   Experiment: 144223 features
   Synthetic:  1928 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


## Summary statistics for: A3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A3/NotGuided/transcripts.gtf

   Experiment: 147191 features
   Synthetic:  2037 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


## Summary statistics for: B1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B1/NotGuided/transcripts.gtf

   Experiment: 108316 features
   Synthetic:  1691 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


## Summary statistics for: B2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B2/NotGuided/transcripts.gtf

   Experiment: 107544 features
   Synthetic:  1616 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


## Summary statistics for: B3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B2/NotGuided/transcripts.gtf

   Experiment: 107544 features
   Synthetic:  1616 features

   Reference:  162 exons
   Reference:  1028 introns

   ***
   *** The following statistics are computed for exact and fuzzy.
   ***
   *** The fuzzy level is 10 nucleotides.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)
   -------------------- Intron level --------------------

   Sensitivity: 0.996032 (0.996032)
   Specificity: 0.996032 (0.996032)
   -------------------- Base level --------------------

   Sensitivity: 1
   Specificity: 1
   -------------------- Intron Chain level --------------------

   Sensitivity: 1 (1)
   Specificity: 1 (1)

   -------------------- Transcript level --------------------

   Sensitivity: 1 (0.993939)
   Specificity: 1 (0.993939)

   Missing exons: 0/872 (0)
   Missing introns: 3/756 (0.00396825)

   Novel exons: 0/872 (0)
   Novel introns: 0/756 (0)



```


\pagebreak


\pagebreak


# TransQuin Expression (Gene)

## Expression summary statistics for: A1

```{ eval=FALSE}

```


\pagebreak


## Expression scatter plot for: A1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(86.9486, 96.745, 3758.17, 1047.79, 55.1694, 37407.9, 3300.12, 622.477, 93.213, 1931.88, 1347.26, 150.979, 15.2912, 32651.3, 72767, 1778.63, 2401.03, 16.2944, 46.4194, 37.6703, 12.6366, 28808.5, 117754, 20.1188, 7491.47, 362.366, 20204.1, 167.12, 18.2308, 667.599, 185.609, 11.608, 8.38523, 1019.99, 49.3218, 347.963, 117035, 5955.88, 15.4836, 149.082, 29.0186, 21487.6, 72072.4, 12446.1, 104.5, 44.0285, 8276.94, 48.9987, 22.042, 26.6032, 6.03984, 23.9704, 15.9987, 1530.13, 36.0393, 39.259, 20.7601, 588.915, 25.2341, 6244.65, 59.6368, 14.562, 255.073, 69.9969, 6775.15, 17.3833, 21.8608, 18.929, 3667.45, 40.9916, 18.488, 20.5209, 20.7893))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: A2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    4.62321 gene

   ***
   *** Detection Limits
   ***

   Break: 3.77655 (R1_62)

   Left:  4.62321 + 0.0227324x (R2 = 0.00360572)
   Right: 2.20463 + 0.990951x (R2 = 0.91596)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.960008
   Slope:       6.60733
   R2:          0.921615
   F-statistic: 834.781
   P-value:     0
   SSM:         3.663e+10, DF: 1
   SSE:         3.11546e+09, DF: 71
   SST:         3.97454e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.925689
   Slope:       0.699133
   R2:          0.8569
   F-statistic: 425.155
   P-value:     0
   SSM:         967.772, DF: 1
   SSE:         161.616, DF: 71
   SST:         1129.39, DF: 72


```


\pagebreak


## Expression scatter plot for: A2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(85.435, 100.905, 3736.98, 1149.2, 59.4362, 38394.7, 3263.84, 683.137, 108.678, 1937.09, 1344.47, 158.068, 20.6353, 32295.2, 74628.3, 1758.79, 2372.79, 22.4385, 40.1739, 45.5986, 12.1864, 28166.8, 121758, 23.2676, 6764.15, 385.518, 19101.5, 180.522, 24.0939, 611.081, 188.008, 9.74278, 9.02527, 988.693, 49.3037, 342.131, 120242, 5175.77, 18.1181, 158.752, 33.4466, 21403.7, 72627.1, 12062.2, 109.784, 42.817, 8092.31, 55.5725, 32.384, 39.6126, 5.6702, 28.9628, 24.6661, 1510.76, 37.5371, 52.0678, 27.9532, 581.666, 35.8495, 6304.42, 66.5508, 20.6696, 257.151, 85.2934, 6899.09, 18.2791, 20.4355, 21.3052, 3849.71, 47.3217, 23.2432, 25.9138, 33.175))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: A3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking

   Experiment:  60569 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    5.25776 gene

   ***
   *** Detection Limits
   ***

   Break: 15.1062 (R1_101)

   Left:  5.25776 + 0.16804x (R2 = 0.139073)
   Right: 1.51144 + 1.05768x (R2 = 0.903529)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.961001
   Slope:       6.31637
   R2:          0.923522
   F-statistic: 857.371
   P-value:     0
   SSM:         3.34749e+10, DF: 1
   SSE:         2.7721e+09, DF: 71
   SST:         3.6247e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.915017
   Slope:       0.666601
   R2:          0.837255
   F-statistic: 365.266
   P-value:     0
   SSM:         879.802, DF: 1
   SSE:         171.015, DF: 71
   SST:         1050.82, DF: 72


```


\pagebreak


## Expression scatter plot for: A3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(84.6473, 111.791, 3773.37, 1055.87, 67.8717, 37398.6, 3123.1, 764.926, 111.389, 1910.16, 1331.34, 151.185, 32.7605, 31802.4, 72780.9, 1796.11, 2620.05, 27.1731, 48.5226, 51.0422, 22.3246, 28284.1, 116089, 31.4486, 7029.58, 363.326, 19387.3, 180.159, 31.3369, 622.701, 192.468, 19.0192, 7.2308, 999.325, 83.9476, 353.492, 113321, 5403.58, 21.2101, 163.727, 48.6509, 21459.4, 69322.3, 12056.9, 117.322, 52.7741, 8141.63, 61.1288, 48.8798, 48.6792, 2.43974, 46.0854, 45.5863, 1508.92, 56.355, 83.8905, 43.7215, 579.118, 54.5987, 6389.2, 74.9619, 27.9453, 315.05, 104.529, 6697.33, 17.0121, 30.9176, 23.5949, 3704.14, 61.0756, 38.8357, 36.411, 38.0141))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking

   Experiment:  60568 gene
   Synthetic:   74 gene

   Reference:   76 gene
   Detected:    5.19759 gene

   ***
   *** Detection Limits
   ***

   Break: 1.88828 (R1_72)

   Left:  5.19759 + 0.42932x (R2 = 0.278885)
   Right: 2.76376 + 0.857773x (R2 = 0.701784)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.87057
   Slope:       3.70819
   R2:          0.757893
   F-statistic: 219.128
   P-value:     0
   SSM:         3.24184e+10, DF: 1
   SSE:         1.0356e+10, DF: 70
   SST:         4.27744e+10, DF: 71

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.84233
   Slope:       0.652163
   R2:          0.709519
   F-statistic: 170.98
   P-value:     0
   SSM:         856.494, DF: 1
   SSE:         350.653, DF: 70
   SST:         1207.15, DF: 71


```


\pagebreak


## Expression scatter plot for: B1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_55','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_66','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 30937.5, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 30937.5, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(23.6283, 18.7275, 2064.3, 62.7222, 31.5313, 56914.9, 1651.54, 61.4493, 100.715, 108.509, 11407.2, 19.8226, 12.6188, 18420.9, 7724.77, 13115.4, 167.17, 58.8514, 88.6405, 14.8761, 10.5637, 115147, 11.2132, 795.714, 361.272, 24725.9, 803.976, 50.1686, 441.211, 1132.63, 29.7886, 19.2736, 5428.77, 42.3412, 411.418, 683.183, 13.0522, 164.454, 25.5413, 10829.3, 76232.4, 1577.17, 105.996, 42.7735, 16.5476, 30.1425, 52.8477, 0.981702, 19.1145, 19.8085, 400.319, 24.5699, 41.2743, 21.6032, 5653.95, 26.8074, 7045.75, 143429, 19.2761, 13.791, 1682.11, 672.188, 6378.33, 6.47327, 53494.9, 4.81958, 175.496, 3871.26, 38.3841, 18.49, 31.0562, 21.9067))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   74 gene

   Reference:   76 gene
   Detected:    5.22081 gene

   ***
   *** Detection Limits
   ***

   Break: 1.88828 (R1_72)

   Left:  5.22081 + 0.303379x (R2 = 0.28516)
   Right: 2.85593 + 0.848121x (R2 = 0.700146)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.866031
   Slope:       3.64019
   R2:          0.750009
   F-statistic: 210.01
   P-value:     0
   SSM:         3.12402e+10, DF: 1
   SSE:         1.04129e+10, DF: 70
   SST:         4.16532e+10, DF: 71

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.836934
   Slope:       0.63197
   R2:          0.700459
   F-statistic: 163.691
   P-value:     0
   SSM:         804.275, DF: 1
   SSE:         343.936, DF: 70
   SST:         1148.21, DF: 71


```


\pagebreak


## Expression scatter plot for: B2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_55','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_66','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 30937.5, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 30937.5, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(21.8571, 18.5253, 2071.22, 69.7603, 35.9613, 57992.6, 1679.03, 56.7917, 99.2867, 110.498, 11327.2, 20.9645, 13.9372, 18049, 7687.01, 12898.9, 167.982, 59.0646, 88.1024, 22.9411, 9.27359, 118102, 13.7815, 779.073, 366.678, 25367.8, 772.579, 49.26, 405.288, 1188.96, 27.2648, 23.4901, 5313.02, 46.5296, 401.238, 641.836, 11.2154, 161.44, 27.7641, 10857.5, 74795.8, 1563.36, 107.952, 46.5558, 20.2159, 34.8663, 59.0324, 3.89162, 22.6932, 19.361, 377.775, 32.4507, 41.3824, 25.3422, 5731.07, 29.3086, 6748.46, 137080, 18.7509, 15.0831, 1550.34, 682.024, 6343.42, 6.73687, 53393.5, 6.6995, 172.249, 4017.25, 37.7848, 25.0608, 31.8361, 25.2375))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   74 gene

   Reference:   76 gene
   Detected:    5.22081 gene

   ***
   *** Detection Limits
   ***

   Break: 1.88828 (R1_72)

   Left:  5.22081 + 0.303379x (R2 = 0.28516)
   Right: 2.85593 + 0.848121x (R2 = 0.700146)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.866031
   Slope:       3.64019
   R2:          0.750009
   F-statistic: 210.01
   P-value:     0
   SSM:         3.12402e+10, DF: 1
   SSE:         1.04129e+10, DF: 70
   SST:         4.16532e+10, DF: 71

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.836934
   Slope:       0.63197
   R2:          0.700459
   F-statistic: 163.691
   P-value:     0
   SSM:         804.275, DF: 1
   SSE:         343.936, DF: 70
   SST:         1148.21, DF: 71


```


\pagebreak


## Expression scatter plot for: B3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 11:38:09.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TG -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -level gene -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/B3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_55','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_66','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 30937.5, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 30937.5, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(21.8571, 18.5253, 2071.22, 69.7603, 35.9613, 57992.6, 1679.03, 56.7917, 99.2867, 110.498, 11327.2, 20.9645, 13.9372, 18049, 7687.01, 12898.9, 167.982, 59.0646, 88.1024, 22.9411, 9.27359, 118102, 13.7815, 779.073, 366.678, 25367.8, 772.579, 49.26, 405.288, 1188.96, 27.2648, 23.4901, 5313.02, 46.5296, 401.238, 641.836, 11.2154, 161.44, 27.7641, 10857.5, 74795.8, 1563.36, 107.952, 46.5558, 20.2159, 34.8663, 59.0324, 3.89162, 22.6932, 19.361, 377.775, 32.4507, 41.3824, 25.3422, 5731.07, 29.3086, 6748.46, 137080, 18.7509, 15.0831, 1550.34, 682.024, 6343.42, 6.73687, 53393.5, 6.6995, 172.249, 4017.25, 37.7848, 25.0608, 31.8361, 25.2375))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


\pagebreak


# TransQuin Expression (Isoform)

## Expression summary statistics for: A1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak

## Minor/Major plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
    #
    # Demostrate how an FPKM table can be used to construct a splicing plot
    #
    
    setwd('/Users/tedwong/Sources/QA/r')
    
    B1 <- read.csv('tests/data/K_562/B1/t_data.ctab', row.names=1, sep='\t')
    B2 <- read.csv('tests/data/K_562/B2/t_data.ctab', row.names=1, sep='\t')
    B3 <- read.csv('tests/data/K_562/B3/t_data.ctab', row.names=1, sep='\t')    

    B1 <- B1[B1$chr=='chrT',]
    B2 <- B2[B2$chr=='chrT',]
    B3 <- B3[B3$chr=='chrT',]
    B1 <- B1[with(B1, order(t_name)),]
    B2 <- B2[with(B2, order(t_name)),]
    B3 <- B3[with(B3, order(t_name)),]    

    data <- data.frame(B1=B1$FPKM, B2=B2$FPKM, B3=B3$FPKM)
    row.names(data) <- B1$t_name

    data <- TransQuin(seqs=row.names(data), B1=data$B1, B2=data$B2, B3=data$B3)
    r <- plotSplice(data)
```

## Expression scatter plot for: A1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: A2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak


## Expression scatter plot for: A2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: A3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak


## Expression scatter plot for: A3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak


## Expression scatter plot for: B1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak


## Expression scatter plot for: B2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression summary statistics for: B3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab

   Experiment:  100769 isoform
   Synthetic:   164 isoform

   Reference:   162 isoform
   Detected:    1.25307 isoform

   ***
   *** Detection Limits
   ***

   Break: 3.5544 (R1_32_1)

   Left:  1.25307 + 0.157653x (R2 = 0.0887965)
   Right: 0.825739 + 1.06985x (R2 = 0.865753)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.951062
   Slope:       5.16984
   R2:          0.904519
   F-statistic: 1373.62
   P-value:     0
   SSM:         7.55327e+10, DF: 1
   SSE:         7.97327e+09, DF: 145
   SST:         8.35059e+10, DF: 146

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.940142
   Slope:       0.820425
   R2:          0.883867
   F-statistic: 1103.57
   P-value:     0
   SSM:         3387.56, DF: 1
   SSE:         445.096, DF: 145
   SST:         3832.65, DF: 146


```


\pagebreak


## Expression scatter plot for: B3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:24:40.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__/TI -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft StringTie -level isoform -ufiles /Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab,/Users/tedwong/Desktop/K_562/StringTie/A1/t_data.ctab 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs     = c('R1_101_1','R1_101_2','R1_102_1','R1_102_2','R1_103_1','R1_103_2','R1_11_1','R1_11_2','R1_12_1','R1_12_2','R1_13_2','R1_14_1','R1_21_1','R1_21_2','R1_22_1','R1_22_2','R1_23_1','R1_23_2','R1_24_1','R1_31_1','R1_31_2','R1_32_1','R1_32_2','R1_33_1','R1_33_2','R1_41_1','R1_41_2','R1_42_1','R1_42_2','R1_51_1','R1_51_2','R1_52_1','R1_52_2','R1_53_1','R1_61_1','R1_61_2','R1_62_1','R1_62_2','R1_63_1','R1_63_2','R1_71_1','R1_71_2','R1_72_1','R1_72_2','R1_81_1','R1_81_2','R1_82_1','R1_82_2','R1_83_1','R1_83_2','R1_91_1','R1_91_2','R1_92_1','R1_92_2','R1_93_1','R1_93_2','R2_105_1','R2_115_1','R2_115_2','R2_116_1','R2_116_2','R2_116_3','R2_117_1','R2_117_3','R2_14_1','R2_14_2','R2_14_3','R2_150_1','R2_150_2','R2_151_1','R2_151_2','R2_151_3','R2_152_1','R2_152_2','R2_153_1','R2_153_2','R2_153_3','R2_154_1','R2_154_2','R2_18_1','R2_18_2','R2_19_1','R2_19_2','R2_1_1','R2_20_1','R2_24_2','R2_26_1','R2_26_2','R2_27_1','R2_27_2','R2_28_1','R2_28_2','R2_28_3','R2_32_1','R2_32_2','R2_32_3','R2_33_1','R2_37_1','R2_37_2','R2_37_3','R2_41_1','R2_42_1','R2_42_2','R2_42_3','R2_45_1','R2_45_2','R2_45_3','R2_45_4','R2_46_1','R2_46_3','R2_47_1','R2_47_2','R2_53_1','R2_53_2','R2_53_3','R2_54_1','R2_54_2','R2_55_2','R2_55_3','R2_57_1','R2_57_2','R2_59_1','R2_59_2','R2_59_3','R2_60_1','R2_60_2','R2_63_1','R2_63_3','R2_65_1','R2_66_1','R2_66_2','R2_67_1','R2_68_1','R2_68_2','R2_6_1','R2_6_2','R2_71_1','R2_71_2','R2_72_1','R2_72_2','R2_73_1','R2_73_2','R2_76_1','R2_76_2','R2_76_3','R2_7_1','R2_7_2'),
                  expected = c(10.0708, 5.0354, 0.8886, 14.2176, 107.422, 859.375, 161.133, 80.5664, 1.7772, 28.4352, 2578.12, 483.398, 3437.5, 27500, 53.7109, 429.688, 0.8886, 14.2176, 483.398, 26.8555, 214.844, 3.5544, 56.8704, 0.0786781, 0.0393391, 454.963, 7279.41, 5156.25, 2578.12, 966.797, 966.797, 0.629425, 0.314713, 120.85, 0.4443, 7.1088, 2.5177, 1.25885, 227.482, 3639.71, 7734.38, 7734.38, 1.25885, 0.629425, 60.4248, 60.4248, 3750, 117.188, 20.1416, 10.0708, 0.0143051, 0.457764, 120.85, 120.85, 30.2124, 30.2124, 0.944138, 80.5664, 40.2832, 0.269754, 0.539507, 1.07901, 30.2124, 30.2124, 2209.82, 4419.64, 8839.29, 58.5938, 1875, 0.269754, 0.539507, 1.07901, 15.1062, 15.1062, 0.134877, 0.269754, 0.539507, 1933.59, 1933.59, 7734.38, 7734.38, 2578.12, 1289.06, 0.944138, 20.1416, 13.4277, 1875, 58.5938, 0.839233, 6.71387, 0.0674384, 0.134877, 0.269754, 0.0674384, 0.134877, 0.269754, 0.0590086, 0.0337192, 0.0674384, 0.134877, 161.133, 0.0337192, 0.0674384, 0.134877, 0.0314713, 0.0629425, 0.125885, 0.25177, 0.0168596, 0.0674384, 80.5664, 40.2832, 0.0168596, 0.0337192, 0.0674384, 322.266, 161.133, 937.5, 30000, 0.4443, 7.1088, 0.0674384, 0.134877, 0.269754, 0.4443, 7.1088, 29.2969, 937.5, 3.77655, 937.5, 30000, 3.77655, 0.114441, 3.66211, 28.4352, 454.963, 0.114441, 3.66211, 0.0157356, 0.0314713, 0.0572205, 1.83105, 0.0084298, 0.0168596, 0.0337192, 29.2969, 937.5),
                  measured = c(31.0537, 14.8166, 3.99894, 55.0263, 285.554, 2099.93, 357.607, 243.951, 0.864198, 33.2104, 13655, 2152.84, 25383.5, 104017, 43.6783, 357.046, 5.07994, 54.8913, 1417.12, 68.3869, 833.828, 6.23008, 78.5313, 1.98259, 2.79967, 1100.59, 20682, 36275.6, 13127, 759.695, 617.179, 4.26125, 0.968966, 28.9282, 1.59312, 13.5461, 1.144, 3.25819, 1192.63, 17952.7, 34344.7, 30124.1, 4.25562, 3.42941, 84.1894, 120.882, 13165.7, 259.377, 52.1735, 28.8282, 1.78034, 3.01341, 253.763, 186.804, 73.6974, 36.3831, 2.04001, 358.996, 239.413, 3.663, 2.87576, 9.88016, 76.7036, 157.674, 8335.2, 24642.2, 43274, 193.636, 4217.88, 0.481615, 0.844077, 1.55965, 47.7706, 23.4573, 1.79312, 1.95146, 2.7487, 6613.41, 5600.6, 13893.7, 27979.1, 6213.78, 1849.94, 3.19921, 44.7132, 22.3072, 5275.27, 365.571, 2.27418, 25.0013, 1.86387, 2.43981, 1.58588, 1.71801, 2.42336, 3.40536, 0.740778, 2.0237, 1.86459, 0.14593, 790.469, 1.76915, 3.25809, 2.92628, 2.46528, 2.98835, 3.25917, 2.82545, 1.75105, 1.00016, 147.722, 305.184, 1.42671, 1.43156, 0.160952, 2686.1, 2006.14, 7524.63, 222838, 1.22333, 31.4029, 0.895378, 0.822505, 2.14064, 2.3159, 33.2361, 0.40669, 4174.12, 12.0058, 7527.22, 137370, 10.7958, 0.258587, 7.75123, 32.8577, 4.11644, 1.50613, 10.5797, 0.826705, 1.65082, 2.16289, 6.36911, 1.62531, 1.14674, 1.87766, 65.0715, 2497.3))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


\pagebreak


# TransQuin Differential

## Differential summary statistics

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv

   Experiment:  60500 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    75 gene

   ***
   *** Detection Limits
   ***

   Absolute:    0.0590086 (attomol/ul) (R2_38)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.707428
   Slope:       0.426606
   R2:          0.500454
   F-statistic: 73.1328
   P-value:     1.29274e-12
   SSM:         302.202, DF: 1
   SSE:         301.653, DF: 73
   SST:         603.855, DF: 74

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.790229
   Slope:       0.648599
   R2:          0.624462
   F-statistic: 121.388
   P-value:     0
   SSM:         183.058, DF: 1
   SSE:         110.087, DF: 73
   SST:         293.145, DF: 74


```


\pagebreak


## ROC plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:34:21.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransDiff -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft DESeq2 -csoft HTSeqCount -level Gene -cfiles /Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA3v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB3v2.htseq.counts.combined -ufiles /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_21','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_55','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_66','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  pval = c(0.0520723, 0.00114072, 6.25e-27, 7.49e-20, 0.175439, 0.59811, 8.56e-10, 0.266285, 1.85e-21, 0.929809, 1.22e-205, 1.55138e-247, 0.000216566, 0.954847, 4.63e-169, 0, 0, 2.19e-51, 0.000455818, 0.329628, 0.280646, 0.869632, 4.01e-227, 0.923951, 0.601793, 2.48e-210, 0.897165, 0.00420847, 2.51e-15, 0.0706309, 0.00544588, 6.46e-36, 0.0345875, 0.468073, 6.22e-84, 0.576496, 0.425398, 1.54e-49, 4.36e-148, 0.438678, 0.555108, 0.448449, 2.05e-31, 0.134721, 5.66999e-297, 0.846458, 0.347465, 0, 0.0112088, 0.978386, 0.0444392, 0.459634, 0.551343, 1.1e-48, 0.85096, 0.308861, 0.63098, 1.75484e-248, 0.755296, 0.0864418, 5.19e-123, 0.0117735, 0.218838, 1.26e-18, 3.36e-21, 8.73e-40, 0.194597, 7.91e-198, 0.294138, 7.73e-14, 0.407513, 0.358588, 0.319528, 0.333173, 0.11141))

# Classify whether they're TP or FP
data <- transClassify(data, lvl='gene')

plotROC(data)

```


\pagebreak


## MA plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:34:22.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransDiff -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft DESeq2 -csoft HTSeqCount -level Gene -cfiles /Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA3v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB3v2.htseq.counts.combined -ufiles /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv 
#

library(Anaquin)

# Read the differential results
data <- read.csv('/Users/tedwong/Sources/QA/__temp__/TransDiffs_diffs.csv', row.name=1)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs=row.names(data), baseMean=data$baseMean, log2FoldChange=data$lfc, lfcSE=data$lfcSE, pvalue=data$pval, expected.LFC=data$elfc)

plotMA(data, lvl='gene', shouldEndo=TRUE)

```


\pagebreak


## LODR plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 22-01-2016 12:34:22.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransDiff -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft DESeq2 -csoft HTSeqCount -level Gene -cfiles /Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA3v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB3v2.htseq.counts.combined -ufiles /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv 
#

library(Anaquin)

# Read the differential results
data <- read.csv('/Users/tedwong/Sources/QA/__temp__/TransDiffs_diffs.csv', row.name=1)

# Create a TransQuin data set for Anaquin
data <- TransQuin(seqs=row.names(data), baseMean=data$baseMean, log2FoldChange=data$lfc, lfcSE=data$lfcSE, pvalue=data$pval, expected.LFC=data$elfc)

# Change to your chooden FDR rate
chosenFDR <- 0.1

plotLODR(data, choseFDR=0.1, shouldTable=FALSE, lvl='gene', shouldBand=FALSE, yBreaks=c(1e-300, 1e-200, 1e-100, 1e-10, 1.00), locBand='local')

```


\pagebreak

