---
title: "Anaquin: TransQuin Report"
header-includes: \usepackage{graphicx}
output: 
    pdf_document:
        keep_tex: true
        toc: yes
        toc_depth: 2
---


\pagebreak


# TransQuin Alignment

## Alignment summary statistics for: A1

```{ eval=FALSE}
Summary for dataset: A1/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 6635268 (21.5363%) reads
   Synthetic:  24174356 (78.4637%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      59421555 exons
   Query:      24855847 introns
   Query:      163383 bases

   Dilution:   0.784637

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.994958
   Specificity: 0.973343
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.991245
   Specificity: 0.821653
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.692468
   Specificity: 0.924613
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.005042
   Intron: 0.008755
   Gene:   0.052632



```


\pagebreak


## Alignment summary statistics for: A2

```{ eval=FALSE}
Summary for dataset: A2/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 5742035 (20.4187%) reads
   Synthetic:  22379420 (79.5813%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      55184443 exons
   Query:      24887760 introns
   Query:      163986 bases

   Dilution:   0.795813

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.996639
   Specificity: 0.975178
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.991245
   Specificity: 0.763722
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.693059
   Specificity: 0.921999
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.003361
   Intron: 0.008755
   Gene:   0.039474



```


\pagebreak


## Alignment summary statistics for: A3

```{ eval=FALSE}
Summary for dataset: A3/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 5803436 (21.0203%) reads
   Synthetic:  21805238 (78.9797%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      53748651 exons
   Query:      22830699 introns
   Query:      166153 bases

   Dilution:   0.789797

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.996639
   Specificity: 0.974466
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.992218
   Specificity: 0.811167
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.695741
   Specificity: 0.913495
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.003361
   Intron: 0.007782
   Gene:   0.039474



```


\pagebreak


## Alignment summary statistics for: B1

```{ eval=FALSE}
Summary for dataset: B1/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 3564555 (14.1897%) reads
   Synthetic:  21556180 (85.8103%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      57069428 exons
   Query:      26914113 introns
   Query:      158601 bases

   Dilution:   0.858103

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.992437
   Specificity: 0.981463
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.986381
   Specificity: 0.880564
   Detection:   1.88828 (R1_72)

   -------------------- Base level --------------------

   Sensitivity: 0.685079
   Specificity: 0.942327
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.007563
   Intron: 0.013619
   Gene:   0.065789



```


\pagebreak


## Alignment summary statistics for: B2

```{ eval=FALSE}
Summary for dataset: B2/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 3673262 (14.5259%) reads
   Synthetic:  21614486 (85.4741%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      57264502 exons
   Query:      28885125 introns
   Query:      160649 bases

   Dilution:   0.854741

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.994118
   Specificity: 0.981653
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.984436
   Specificity: 0.818059
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.687879
   Specificity: 0.934117
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.005882
   Intron: 0.015564
   Gene:   0.052632



```


\pagebreak


## Alignment summary statistics for: B3

```{ eval=FALSE}
Summary for dataset: B3/TransAlign_summary.stats

   Unmapped:   0 reads
   Experiment: 4705904 (15.6992%) reads
   Synthetic:  25269549 (84.3008%) reads

   Reference:  1190 exons
   Reference:  1028 introns
   Reference:  149219 bases

   Query:      67685272 exons
   Query:      37139930 introns
   Query:      162712 bases

   Dilution:   0.843008

   ***
   *** The following statistics are computed at the exon, intron and base level.
   ***
   *** Exon level is defined by performance per exon. An alignment that
   *** is not mapped entirely within an exon is considered as a FP. The
   *** intron level is similar.
   ***
   *** Base level is defined by performance per nucleotide. A partial
   *** mapped read will have FP and TP.
   ***

   -------------------- Exon level --------------------

   Sensitivity: 0.994958
   Specificity: 0.980831
   Detection:   0.0590086 (R2_33)

   -------------------- Intron level --------------------

   Sensitivity: 0.986381
   Specificity: 0.759454
   Detection:   0.0590086 (R2_33)

   -------------------- Base level --------------------

   Sensitivity: 0.688292
   Specificity: 0.922827
   Detection:   0.0590086 (R2_33)

   -------------------- Undetected --------------------

   Exon:   0.005042
   Intron: 0.013619
   Gene:   0.052632



```


\pagebreak


## Alignment sequin statistics for: A1

```{ eval=FALSE}
Summary for dataset: A1/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.847694	1	0.999322	1	1	0.847694	0.998974
R1_102	0.721248	1	0.996879	1	0.11205	0.721248	0.998544
R1_103	0.628255	1	0.996404	1	0.872516	0.628255	0.994737
R1_11	0.645161	1	0.999251	1	0.0622323	0.645161	0.996885
R1_12	0.590941	1	0.995265	1	0.998603	0.590941	0.996501
R1_13	0.920773	1	0.992209	0.909091	0.939559	0.920773	0.994269
R1_14	1	0.999757	--	--	1	0.995502	1
R1_21	0.630945	1	0.997646	1	0.91334	0.630945	0.942446
R1_22	0.528959	1	0.9983	1	0.0595461	0.528959	0.993203
R1_23	0.608847	1	0.994273	1	0.00824358	0.608847	0.997717
R1_24	1	1	0.995671	1	0.998218	1	0.997168
R1_31	0.688293	1	0.995361	1	0.179726	0.688293	0.994565
R1_32	0.545455	1	0.996002	1	0.997528	0.545455	0.990536
R1_33	0.88324	1	0.994233	1	0.997701	0.88324	0.998997
R1_41	0.78125	1	0.996444	1	0.950421	0.78125	0.994695
R1_42	0.617479	1	0.982642	1	0.89545	0.617479	0.986842
R1_43	0.526611	0.975	0.99293	0.973684	0.96402	0.526611	0.995551
R1_51	0.607666	1	0.996623	1	0.999144	0.607666	0.992634
R1_52	0.6209	1	0.994545	1	1	0.6209	0.970203
R1_53	0.998489	1	0.998138	1	0.567302	0.998489	0.997485
R1_61	0.686144	1	1	1	1	0.686144	1
R1_62	0.776233	1	1	1	1	0.776233	1
R1_63	0.669964	1	0.997411	1	0.987252	0.669964	0.99536
R1_71	0.740968	1	0.989782	1	0.996867	0.740968	0.985685
R1_72	0.605418	1	1	1	1	0.605418	1
R1_73	0.731952	1	0.990579	1	0.988759	0.731952	0.996297
R1_81	0.746457	1	0.996933	1	0.996006	0.746457	0.996322
R1_82	0.587741	1	0.991724	1	0.993417	0.587741	0.993612
R1_83	0.644813	1	0.998035	1	0.997838	0.644813	0.997018
R1_91	0.683772	1	1	1	1	0.683772	1
R1_92	0.777496	1	0.990108	1	0.997403	0.777496	0.995127
R1_93	0.625086	1	0.995626	1	0.997761	0.625086	0.997251
R2_1	1	1	--	--	0.996985	1	0.996985
R2_105	1	0.942857	--	--	0.874667	0.99696	0.874667
R2_115	0.837412	1	0.992879	1	0.996245	0.837412	0.922708
R2_116	0.584726	1	0.996656	1	1	0.584726	0.997557
R2_117	0.710041	1	0.997963	1	0.986627	0.710041	0.99569
R2_14	0.719569	1	0.986972	1	0.948655	0.719569	0.991507
R2_150	0.834679	1	0.996211	1	0.599452	0.834679	0.991777
R2_151	0.573928	1	0.97485	1	0.00142271	0.573928	0.994693
R2_152	0.60288	1	0.997882	1	0.017786	0.60288	0.99798
R2_153	0.671455	1	0.995719	1	1	0.671455	0.99909
R2_154	0.855658	1	0.988294	1	0.641379	0.855658	0.993964
R2_18	0.628326	1	0.984616	1	0.92933	0.628326	0.985276
R2_19	0.811493	1	0.992047	1	0.947606	0.811493	0.998147
R2_20	0.732411	1	0.998754	1	1	0.732411	0.997905
R2_24	0.586003	0.979592	0.994692	0.957447	1	0.586003	0.998491
R2_26	0.916294	1	0.995692	1	0.985066	0.916294	0.997941
R2_27	0.754464	1	0.998102	1	1	0.754464	0.999212
R2_28	0.685301	1	1	1	1	0.685301	1
R2_32	0.669025	1	0.986915	1	1	0.669025	0.998623
R2_33	0.879859	1	1	1	0.000207469	0.879859	1
R2_37	0.724904	1	0.994058	1	0.994505	0.724904	0.999687
R2_38	0.375412	0.888889	0.987441	0.857143	1	0.375412	0.9965
R2_41	0.838456	1	0.99339	1	0.995872	0.838456	0.967423
R2_42	0.883333	1	0.997452	1	1	0.883333	0.999123
R2_45	0.464239	1	1	1	1	0.464239	1
R2_46	0.706554	1	0.988095	1	0.984375	0.706554	0.96668
R2_47	0.888126	1	0.996247	1	0.997381	0.888126	0.998676
R2_53	0.389676	1	0.99433	1	0.990033	0.389676	0.997705
R2_54	0.864703	1	0.992248	1	0.857243	0.864703	0.997446
R2_55	0.880111	1	0.990548	1	0.850544	0.880111	0.988327
R2_57	0.64724	1	0.996649	1	0.00781695	0.64724	0.998809
R2_59	0.594477	1	0.997976	1	0.995441	0.594477	0.999728
R2_6	0.702025	1	0.995992	1	0.989247	0.702025	0.98913
R2_60	0.606114	1	0.9967	1	0.00853443	0.606114	0.997076
R2_63	0.883925	1	0.988671	1	0.98321	0.883925	0.996051
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.529254	1	0.992772	1	0.999104	0.529254	0.985015
R2_67	0.959847	1	1	1	1	0.959847	1
R2_68	0.699062	1	0.99951	1	1	0.699062	0.999681
R2_7	0.76132	1	0.998183	1	0.999392	0.76132	0.996124
R2_71	0.822193	1	0.991979	1	1	0.822193	0.999188
R2_72	0.343281	0.866667	0.987707	0.727273	1	0.343281	0.998704
R2_73	0.931968	1	0.997207	1	0.998643	0.931968	0.999512
R2_76	0.472826	1	0.997807	1	1	0.472826	0.999234

```


\pagebreak


## Alignment sequin statistics for: A2

```{ eval=FALSE}
Summary for dataset: A2/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.842472	1	0.997743	1	0.994398	0.842472	0.998968
R1_102	0.72195	1	0.996092	1	0.00745621	0.72195	0.998061
R1_103	0.629363	1	0.996332	1	0.190876	0.629363	0.995182
R1_11	0.645833	1	0.999611	1	0.0609563	0.645833	0.99896
R1_12	0.591632	1	0.996195	1	0.999449	0.591632	0.997668
R1_13	0.920773	1	0.99341	0.909091	0.689422	0.920773	0.994676
R1_14	1	0.999704	--	--	1	0.994012	1
R1_21	0.630945	1	0.99783	1	0.94016	0.630945	0.940754
R1_22	0.528959	1	0.99921	1	0.0886522	0.528959	0.995741
R1_23	0.605016	1	0.990962	1	0.0131643	0.605016	0.996558
R1_24	1	1	0.995808	1	0.994776	1	0.99782
R1_31	0.689234	1	0.994005	1	0.99822	0.689234	0.995248
R1_32	0.545455	1	0.996609	1	1	0.545455	0.99472
R1_33	0.911913	1	0.997009	1	1	0.911913	0.999028
R1_41	0.78125	1	0.996615	1	0.984857	0.78125	0.995575
R1_42	0.617479	1	0.983941	1	0.918226	0.617479	0.990854
R1_43	0.526611	0.975	0.992939	0.973684	0.970546	0.526611	0.995551
R1_51	0.607103	1	0.995561	1	0.995538	0.607103	0.993542
R1_52	0.623951	1	0.986547	1	0.993277	0.623951	0.970344
R1_53	0.997986	1	0.995796	1	0.998658	0.997986	0.996982
R1_61	0.687813	1	0.996259	1	0.00393677	0.687813	0.997579
R1_62	0.766119	1	0.990741	1	0.00177231	0.766119	0.999176
R1_63	0.670224	1	0.997749	1	0.982678	0.670224	0.994977
R1_71	0.740968	1	0.990143	1	0.965202	0.740968	0.986694
R1_72	0.6	1	1	1	0.000771367	0.6	1
R1_73	0.731952	1	0.990777	0.96	0.594568	0.731952	0.996014
R1_81	0.74685	1	0.997109	1	0.994664	0.74685	0.996847
R1_82	0.587741	1	0.989888	1	0.996501	0.587741	0.993612
R1_83	0.644427	1	0.998518	1	0.998462	0.644427	0.996422
R1_91	0.68762	1	1	1	1	0.68762	1
R1_92	0.779611	1	0.989036	1	0.997001	0.779611	0.99514
R1_93	0.625775	1	0.996087	1	0.997967	0.625775	0.997254
R2_1	1	1	--	--	0.990955	1	0.990955
R2_105	1	1	--	--	0.690667	1	0.690667
R2_115	0.840131	1	0.993099	1	0.995773	0.840131	0.992293
R2_116	0.52506	1	0.995198	1	1	0.52506	0.998185
R2_117	0.711066	1	0.997799	1	0.986683	0.711066	0.995696
R2_14	0.719569	1	0.987726	1	0.762312	0.719569	0.990981
R2_150	0.835448	1	0.997498	1	0.637765	0.835448	0.991785
R2_151	0.559055	1	0.993207	1	0.00185397	0.559055	0.999218
R2_152	0.602636	1	0.998542	1	0.0249012	0.602636	0.998383
R2_153	0.679707	1	1	1	1	0.679707	1
R2_154	0.855658	1	0.991999	1	0.752922	0.855658	0.994631
R2_18	0.628326	1	0.98195	1	0.952684	0.628326	0.986486
R2_19	0.811493	1	0.991557	1	0.412105	0.811493	0.99829
R2_20	0.732795	1	0.9974	1	0.997538	0.732795	0.997906
R2_24	0.585856	0.979592	0.991346	0.957447	1	0.585856	0.997737
R2_26	0.916638	1	0.995474	1	0.994598	0.916638	0.997568
R2_27	0.752381	1	0.996593	1	1	0.752381	0.999209
R2_28	0.686747	1	0.992133	1	1	0.686747	0.979381
R2_32	0.671486	1	0.988811	1	0.995984	0.671486	0.997715
R2_33	0.922261	1	1	1	1	0.922261	1
R2_37	0.725357	1	0.994978	1	1	0.725357	0.996265
R2_38	0.394199	1	0.989486	1	0.976667	0.394199	0.998331
R2_41	0.839171	1	0.994146	1	0.995818	0.839171	0.990299
R2_42	0.871318	1	0.998773	1	0.996774	0.871318	0.999555
R2_45	0.464239	1	0.99862	1	0.998657	0.464239	0.999378
R2_46	0.705051	1	0.994681	1	1	0.705051	0.967409
R2_47	0.888126	1	0.99716	1	0.997684	0.888126	0.997795
R2_53	0.39093	1	0.998503	1	1	0.39093	0.999542
R2_54	0.864703	1	0.992071	1	0.99359	0.864703	0.997446
R2_55	0.880111	1	0.990367	1	0.877687	0.880111	0.987558
R2_57	0.659591	1	0.997113	1	0.00776074	0.659591	0.93286
R2_59	0.60449	1	0.999487	1	1	0.60449	0.999733
R2_6	0.693346	1	0.997899	1	1	0.693346	0.997226
R2_60	0.604692	1	0.999074	1	0.00876215	0.604692	0.998239
R2_63	0.86509	1	0.986341	0.666667	0.984241	0.86509	0.996971
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.529254	1	0.993527	1	0.998951	0.529254	0.986987
R2_67	0.988528	1	1	1	1	0.988528	1
R2_68	0.717828	1	0.99418	1	1	0.717828	0.999689
R2_7	0.76132	1	0.997537	1	0.998989	0.76132	0.996124
R2_71	0.808155	1	0.995012	1	0.993478	0.808155	0.998348
R2_72	0.341203	0.866667	0.991131	0.727273	1	0.341203	0.99913
R2_73	0.950171	1	0.994921	1	0.99508	0.950171	0.999282
R2_76	0.475725	1	0.996596	1	1	0.475725	0.998479

```


\pagebreak


## Alignment sequin statistics for: A3

```{ eval=FALSE}
Summary for dataset: A3/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.843342	1	0.998545	1	1	0.843342	0.998969
R1_102	0.724404	1	0.996659	1	0.0275419	0.724404	0.998068
R1_103	0.629086	1	0.99598	1	0.466926	0.629086	0.99518
R1_11	0.646505	1	0.999213	1	0.0613864	0.646505	0.998962
R1_12	0.591978	1	0.996492	1	0.999063	0.591978	0.998833
R1_13	0.921152	1	0.992714	0.909091	0.927537	0.921152	0.994272
R1_14	1	0.999661	--	--	1	0.996997	1
R1_21	0.630945	1	0.997923	1	0.932742	0.630945	0.941599
R1_22	0.528959	1	0.999234	1	0.0715257	0.528959	0.99659
R1_23	0.610589	1	0.991844	1	0.00939993	0.610589	0.996023
R1_24	1	1	0.995729	1	0.997156	1	0.997168
R1_31	0.688764	1	0.993699	1	0.998371	0.688764	0.995245
R1_32	0.545455	1	0.996925	1	0.997249	0.545455	0.992624
R1_33	0.911913	1	0.995448	1	0.998975	0.911913	0.998705
R1_41	0.78125	1	0.996329	1	0.949202	0.78125	0.995575
R1_42	0.617479	1	0.986143	1	0.911049	0.617479	0.990854
R1_43	0.526611	0.975	0.993215	0.973684	0.971656	0.526611	0.995892
R1_51	0.607103	1	0.996125	1	0.993582	0.607103	0.993542
R1_52	0.624333	1	0.991891	1	0.996644	0.624333	0.968639
R1_53	0.999496	1	0.995838	1	0.894373	0.999496	0.997487
R1_61	0.705342	1	0.996028	1	0.030201	0.705342	0.998818
R1_62	0.774968	1	0.996558	1	0.0173625	0.774968	0.999185
R1_63	0.669703	1	0.997245	1	0.984976	0.669703	0.995743
R1_71	0.740968	1	0.993089	1	0.991204	0.740968	0.985685
R1_72	0.604966	1	1	1	0.0175711	0.604966	1
R1_73	0.731534	1	0.986479	0.96	0.869319	0.731534	0.996295
R1_81	0.746457	1	0.996425	1	0.258511	0.746457	0.996845
R1_82	0.587741	1	0.995206	1	0.978119	0.587741	0.993612
R1_83	0.644427	1	0.99821	1	0.997789	0.644427	0.998805
R1_91	0.681206	1	1	1	1	0.681206	1
R1_92	0.774535	1	0.988785	1	0.997123	0.774535	0.99565
R1_93	0.625775	1	0.997114	1	0.998507	0.625775	0.997254
R2_1	1	1	--	--	0.99598	1	0.99598
R2_105	1	0.965517	--	--	0.930667	0.994302	0.930667
R2_115	0.837412	1	0.99224	1	0.997013	0.837412	0.922708
R2_116	0.582339	1	0.996341	1	1	0.582339	0.997547
R2_117	0.710041	1	0.99923	1	0.998468	0.710041	0.997122
R2_14	0.719569	1	0.987344	1	0.913125	0.719569	0.991507
R2_150	0.835832	1	0.996781	1	0.569783	0.835832	0.996334
R2_151	0.588364	1	0.990961	1	0.00205912	0.588364	0.997035
R2_152	0.605321	1	0.998972	1	0.0178198	0.605321	0.997586
R2_153	0.684902	1	0.995235	1	0.998407	0.684902	0.998663
R2_154	0.855658	1	0.988085	1	0.643553	0.855658	0.994631
R2_18	0.628326	1	0.98154	1	0.932426	0.628326	0.986486
R2_19	0.811493	1	0.990813	1	0.706767	0.811493	0.99829
R2_20	0.733948	1	0.998507	1	0.998241	0.733948	0.996347
R2_24	0.586594	0.979592	0.9931	0.957447	1	0.586594	0.998743
R2_26	0.916294	1	0.995585	1	0.994024	0.916294	0.997754
R2_27	0.755952	1	0.995752	1	1	0.755952	0.998035
R2_28	0.689639	1	0.998662	1	1	0.689639	0.980137
R2_32	0.671486	1	0.991023	1	1	0.671486	0.998171
R2_33	0.961131	1	0.916667	1	1	0.961131	0.992701
R2_37	0.727623	1	0.994605	1	0.997226	0.727623	0.997515
R2_38	0.398484	1	0.99778	1	1	0.398484	0.998348
R2_41	0.839171	1	0.993471	1	0.99638	0.839171	0.968247
R2_42	0.884884	1	0.99882	1	1	0.884884	0.949667
R2_45	0.464384	1	0.99941	1	0.999372	0.464384	0.999067
R2_46	0.719784	1	0.992964	1	1	0.719784	0.967664
R2_47	0.888714	1	0.996164	1	0.869457	0.888714	0.998456
R2_53	0.391647	1	0.996204	1	0.995389	0.391647	0.997262
R2_54	0.864703	1	0.992777	1	0.996222	0.864703	0.997446
R2_55	0.880804	1	0.990786	1	0.84464	0.880804	0.988336
R2_57	0.661135	1	0.997049	1	0.00967222	0.661135	0.99767
R2_59	0.592539	1	0.996507	1	0.997951	0.592539	0.998639
R2_6	0.700096	1	0.996782	1	0.996937	0.700096	0.995885
R2_60	0.605759	1	0.998467	1	0.0137983	0.605759	0.998243
R2_63	0.884363	1	0.986952	1	0.981729	0.884363	0.997037
R2_65	1	1	--	--	0.994493	1	0.994493
R2_66	0.529254	1	0.99313	1	0.998672	0.529254	0.985015
R2_67	0.971319	1	1	1	1	0.971319	1
R2_68	0.717828	1	0.996276	1	0.997953	0.717828	0.999378
R2_7	0.76132	1	0.997977	1	0.999136	0.76132	0.996124
R2_71	0.916444	1	0.991426	1	0.989924	0.916444	0.998543
R2_72	0.347439	0.866667	0.992315	0.727273	1	0.347439	0.997017
R2_73	0.928328	1	0.996213	1	0.995682	0.928328	0.998776
R2_76	0.473551	1	0.996066	1	1	0.473551	0.998472

```


\pagebreak


## Alignment sequin statistics for: B1

```{ eval=FALSE}
Summary for dataset: B1/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.798085	1	0.996503	1	1	0.798085	0.998911
R1_102	0.719144	1	0.994911	1	0.0477612	0.719144	0.998054
R1_103	0.627978	1	0.992949	1	0.877408	0.627978	0.996046
R1_11	0.644489	1	1	1	0.0383237	0.644489	1
R1_12	0.570194	1	0.993312	1	0.994968	0.570194	0.998184
R1_13	0.921152	1	0.988876	0.909091	0.986406	0.921152	0.993865
R1_14	1	0.999785	--	--	0.998494	0.996992	0.998494
R1_21	0.630945	1	0.997942	1	0.922412	0.630945	0.942446
R1_22	0.525339	1	1	1	0.00512173	0.525339	1
R1_23	0.609892	1	0.990422	1	0.00775832	0.609892	0.996585
R1_24	0.999782	1	0.997355	1	0.99911	0.999782	0.997819
R1_31	0.688764	1	0.993387	1	0.997399	0.688764	0.991875
R1_32	0.541401	1	1	1	1	0.541401	1
R1_33	0.901271	1	0.996101	1	1	0.901271	0.999017
R1_41	0.78125	1	0.999389	1	0.975044	0.78125	0.996457
R1_42	0.617479	1	0.992588	1	0.880227	0.617479	0.990854
R1_43	0.52643	0.975	0.992495	0.973684	0.701694	0.52643	0.994868
R1_51	0.604848	1	0.998724	1	0.341777	0.604848	0.996286
R1_52	0.622807	1	0.994452	1	0.996778	0.622807	0.998166
R1_53	0.998489	1	0.996504	1	0.186155	0.998489	0.996482
R1_61	0.69783	1	1	1	0.893617	0.69783	1
R1_62	0.70354	1	1	1	0.855072	0.70354	1
R1_63	0.670484	1	0.997312	1	0.942241	0.670484	0.995363
R1_71	0.740968	1	0.98966	1	0.969902	0.740968	0.986694
R1_72	0.578781	1	1	1	0.777778	0.578781	1
R1_73	0.731325	1	0.99004	0.96	0.996528	0.731325	0.997432
R1_81	0.747244	1	0.996751	1	0.992513	0.747244	0.997897
R1_82	0.587741	1	0.993087	1	0.988905	0.587741	0.991501
R1_83	0.647898	1	0.997868	1	0.99442	0.647898	0.998217
R1_91	0.682489	1	1	1	1	0.682489	1
R1_92	0.777073	1	0.989744	1	0.997185	0.777073	0.996204
R1_93	0.625431	1	0.997054	1	0.997791	0.625431	0.996705
R2_1	1	1	--	--	0.992965	1	0.992965
R2_105	1	0.985507	--	--	0.706667	0.992509	0.706667
R2_115	0.840674	1	0.991763	1	0.991845	0.840674	0.922985
R2_116	0.512172	1	0.996177	1	1	0.512172	0.999069
R2_117	0.711578	1	0.996356	1	0.996453	0.711578	0.994986
R2_14	0.719569	1	0.982878	1	0.916738	0.719569	0.991507
R2_150	0.832372	1	0.9986	1	0.130011	0.832372	0.997696
R2_151	0.549869	1	0.991817	1	0.00119505	0.549869	0.999205
R2_152	0.603856	1	0.997347	1	0.0157253	0.603856	0.998789
R2_153	0.680012	1	0.991903	1	1	0.680012	0.988011
R2_154	0.855081	1	0.988127	1	0.451182	0.855081	0.994627
R2_18	0.628326	1	0.984865	1	0.925924	0.628326	0.986486
R2_19	0.811146	1	0.994538	1	0.821059	0.811146	0.999144
R2_20	0.732795	1	0.998868	1	0.999023	0.732795	0.998952
R2_24	0.576259	0.979592	0.99326	0.957447	1	0.576259	0.998721
R2_26	0.916638	1	0.993005	1	0.94618	0.916638	0.997195
R2_27	0.752976	1	0.990885	1	1	0.752976	0.99921
R2_28	0.684578	1	1	1	1	0.684578	1
R2_32	0.669333	1	0.994737	1	0.997678	0.669333	0.998165
R2_33	0.441696	0.5	1	0	nan	0.441696	1
R2_37	0.716293	1	0.997416	1	1	0.716293	0.999684
R2_38	0.380686	1	0.991424	1	0.967742	0.380686	0.998271
R2_41	0.838099	1	0.994207	1	0.993615	0.838099	0.990287
R2_42	0.831008	1	0.997712	1	1	0.831008	0.999534
R2_45	0.46395	1	1	1	1	0.46395	1
R2_46	0.709561	1	0.992042	1	1	0.709561	0.998308
R2_47	0.888911	1	0.996396	1	0.986294	0.888911	0.997797
R2_53	0.389317	1	0.993598	1	0.991416	0.389317	0.99908
R2_54	0.864703	1	0.991482	1	0.994474	0.864703	0.997446
R2_55	0.880804	1	0.981805	1	0.997933	0.880804	0.99142
R2_57	0.64917	1	0.99569	1	0.0158003	0.64917	0.998812
R2_59	0.580426	1	0.997258	1	0.994024	0.580426	0.999722
R2_6	0.702989	1	0.996508	1	1	0.702989	0.990489
R2_60	0.608603	1	0.997788	1	0.354215	0.608603	0.997088
R2_63	0.886991	1	0.999101	1	0.99974	0.886991	0.997046
R2_65	1	1	--	--	0.993392	1	0.993392
R2_66	0.528717	1	0.999688	1	0.999066	0.528717	0.987964
R2_67	0.957935	1	1	1	1	0.957935	1
R2_68	0.569258	0.846154	0.996355	0.833333	0.996273	0.569258	0.998824
R2_7	0.76132	1	0.998103	1	0.999407	0.76132	0.996124
R2_71	0.779412	1	0.978	1	0.975352	0.779412	0.998288
R2_72	0.327246	0.866667	0.986853	0.727273	1	0.327246	0.998188
R2_73	0.944027	1	0.996729	0.952381	0.998051	0.944027	0.999037
R2_76	0.463043	1	0.996047	1	1	0.463043	0.999218

```


\pagebreak


## Alignment sequin statistics for: B2

```{ eval=FALSE}
Summary for dataset: B2/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.833768	1	1	1	1	0.833768	1
R1_102	0.719495	1	0.987047	1	0.00176159	0.719495	0.99854
R1_103	0.627701	1	0.992816	1	0.207487	0.627701	0.995169
R1_11	0.644489	1	1	1	0.0448018	0.644489	1
R1_12	0.583333	1	0.99528	1	0.99734	0.583333	0.998225
R1_13	0.920773	1	0.995253	0.909091	0.780182	0.920773	0.993863
R1_14	1	0.999648	--	--	1	0.995502	1
R1_21	0.630945	1	0.997868	1	0.894799	0.630945	0.942446
R1_22	0.527149	1	0.997019	1	0.00852348	0.527149	0.998286
R1_23	0.61024	1	0.993084	1	0.015077	0.61024	0.998291
R1_24	1	1	0.996884	1	0.997909	1	0.998473
R1_31	0.689234	1	0.992408	1	0.99621	0.689234	0.991881
R1_32	0.540822	1	1	1	1	0.540822	1
R1_33	0.880579	1	1	0.96	1	0.880579	1
R1_41	0.78125	1	0.999422	1	0.887919	0.78125	0.993816
R1_42	0.617479	1	0.993277	1	0.891905	0.617479	0.993884
R1_43	0.526249	0.975	0.991493	0.973684	0.720085	0.526249	0.994867
R1_51	0.603157	1	0.995285	1	0.0946672	0.603157	0.997204
R1_52	0.6209	1	0.993337	1	0.998695	0.6209	0.970203
R1_53	0.998489	1	0.996715	1	0.0428192	0.998489	0.996983
R1_61	0.689482	1	1	1	0.00220526	0.689482	1
R1_62	0.774968	1	1	1	0.000468596	0.774968	1
R1_63	0.670484	1	0.997277	1	0.87543	0.670484	0.995363
R1_71	0.740968	1	0.992827	1	0.949823	0.740968	0.983673
R1_72	0.607223	1	1	1	0.00092004	0.607223	1
R1_73	0.731743	1	0.991638	0.96	0.0932937	0.731743	0.996864
R1_81	0.748031	1	0.997098	1	0.04424	0.748031	0.996329
R1_82	0.587741	1	0.99522	1	0.73257	0.587741	0.993612
R1_83	0.647513	1	0.998545	1	0.997938	0.647513	0.998216
R1_91	0.676716	1	0.996396	1	0.989189	0.676716	0.999053
R1_92	0.777496	1	0.989192	1	0.997857	0.777496	0.995127
R1_93	0.62612	1	0.996901	1	0.996568	0.62612	0.995616
R2_1	1	1	--	--	0.99397	1	0.99397
R2_105	1	0.975904	--	--	0.874667	0.993939	0.874667
R2_115	0.838499	1	0.993517	1	0.991845	0.838499	0.922801
R2_116	0.57136	1	0.991803	1	1	0.57136	0.9975
R2_117	0.711066	1	0.997246	1	0.99467	0.711066	0.994982
R2_14	0.719569	1	0.983698	1	0.932459	0.719569	0.991507
R2_150	0.83391	1	0.998194	1	0.214306	0.83391	0.9977
R2_151	0.568241	1	0.982484	1	0.00243339	0.568241	0.997696
R2_152	0.603124	1	0.998362	1	0.0300802	0.603124	0.997981
R2_153	0.671149	1	0.99501	1	1	0.671149	0.987854
R2_154	0.855658	1	0.987492	1	0.600748	0.855658	0.993964
R2_18	0.628326	1	0.984022	1	0.991925	0.628326	0.986486
R2_19	0.811378	1	0.996101	1	0.141851	0.811378	0.999001
R2_20	0.734333	1	0.996728	1	0.996232	0.734333	0.995829
R2_24	0.585856	0.979592	0.992486	0.957447	1	0.585856	0.997988
R2_26	0.916638	1	0.993025	1	0.787821	0.916638	0.997568
R2_27	0.75625	1	0.99559	1	1	0.75625	0.999607
R2_28	0.688434	1	1	1	1	0.688434	1
R2_32	0.66964	1	0.994642	1	1	0.66964	0.989545
R2_33	0.897527	1	1	1	1	0.897527	1
R2_37	0.722184	1	0.987135	0.888889	0.996283	0.722184	0.998434
R2_38	0.378378	1	0.965649	1	1	0.378378	0.998261
R2_41	0.837026	1	0.994549	1	0.997638	0.837026	0.99702
R2_42	0.865891	1	1	1	1	0.865891	1
R2_45	0.463083	1	0.999771	1	1	0.463083	0.999688
R2_46	0.705352	1	1	1	1	0.705352	1
R2_47	0.889107	1	0.996018	1	0.984317	0.889107	0.998017
R2_53	0.389496	1	0.998159	1	1	0.389496	0.99908
R2_54	0.864703	1	0.992218	1	0.997733	0.864703	0.997191
R2_55	0.880804	1	0.981264	1	0.998791	0.880804	0.990647
R2_57	0.624083	1	0.995	0.941176	0.0127108	0.624083	0.998765
R2_59	0.57655	1	1	1	1	0.57655	1
R2_6	0.702025	1	0.995867	1	1	0.702025	0.98913
R2_60	0.608603	1	0.99745	1	0.370078	0.608603	0.996508
R2_63	0.886991	1	0.998516	1	0.999484	0.886991	0.996555
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.527107	1	0.999693	1	0.998859	0.527107	0.987928
R2_67	0.944551	1	1	1	1	0.944551	1
R2_68	0.596738	0.884615	0.996294	0.833333	0.996175	0.596738	0.998878
R2_7	0.76132	1	0.998123	1	0.998683	0.76132	0.995573
R2_71	0.669118	1	0.996303	1	1	0.669118	0.999002
R2_72	0.341203	0.866667	0.987028	0.727273	1	0.341203	0.998696
R2_73	0.929693	1	0.996043	1	0.999042	0.929693	0.999511
R2_76	0.476449	1	0.99776	1	1	0.476449	0.99924

```


\pagebreak


## Alignment sequin statistics for: B3

```{ eval=FALSE}
Summary for dataset: B3/TransAlign_quins.stats

ID	Covered	Sensitivity (Exon)	Specificity (Exon)	Sensitivity (Intron)	Specificity (Intron)	Sensitivity (Base)	Specificity (Base)
R1_101	0.845953	1	1	1	1	0.845953	1
R1_102	0.709327	1	0.986618	1	0.0102989	0.709327	0.998519
R1_103	0.627978	1	0.992584	1	0.632425	0.627978	0.995608
R1_11	0.626344	1	0.998551	1	0.037669	0.626344	0.995726
R1_12	0.590941	1	0.99219	1	0.994792	0.590941	0.998248
R1_13	0.920015	1	0.98925	0.909091	0.951455	0.920015	0.994672
R1_14	1	0.999485	--	--	1	0.989568	1
R1_21	0.630945	1	0.998002	1	0.797785	0.630945	0.940754
R1_22	0.527149	1	0.998893	1	0.00177882	0.527149	0.996578
R1_23	0.609544	1	0.993163	1	0.00284772	0.609544	0.997719
R1_24	0.999563	1	0.996226	1	0.998971	0.999563	0.998037
R1_31	0.689234	1	0.988055	1	0.98054	0.689234	0.992552
R1_32	0.543717	1	1	1	1	0.543717	1
R1_33	0.903045	1	0.996546	1	1	0.903045	0.999019
R1_41	0.78125	1	0.999615	1	0.973023	0.78125	0.996457
R1_42	0.617479	1	0.993761	1	0.876575	0.617479	0.991862
R1_43	0.52643	0.975	0.992023	0.973684	0.491409	0.52643	0.994868
R1_51	0.604284	1	0.996424	1	0.331359	0.604284	0.996283
R1_52	0.625095	1	0.992522	1	0.998908	0.625095	0.996353
R1_53	0.998489	1	0.996255	1	0.207714	0.998489	0.996983
R1_61	0.686978	1	1	1	0.0155595	0.686978	1
R1_62	0.771176	1	1	1	0.00775528	0.771176	1
R1_63	0.670484	1	0.997669	1	0.87759	0.670484	0.994595
R1_71	0.740968	1	0.990145	1	0.965978	0.740968	0.986694
R1_72	0.588713	1	1	1	0.00389273	0.588713	1
R1_73	0.730906	1	0.991228	1	0.395113	0.730906	0.99743
R1_81	0.748425	1	0.996473	1	0.206774	0.748425	0.995809
R1_82	0.587741	1	0.993872	1	0.997645	0.587741	0.994318
R1_83	0.645584	1	0.998169	1	0.998138	0.645584	0.997616
R1_91	0.679923	1	0.998533	1	0.995893	0.679923	0.999057
R1_92	0.776227	1	0.989358	1	0.997596	0.776227	0.995119
R1_93	0.625086	1	0.996451	1	0.997744	0.625086	0.994518
R2_1	1	1	--	--	0.994975	1	0.994975
R2_105	1	0.963855	--	--	0.984	0.994609	0.984
R2_115	0.841218	1	0.993299	1	0.992553	0.841218	0.923031
R2_116	0.518854	1	0.997216	1	1	0.518854	0.999081
R2_117	0.711066	1	0.998624	1	0.998808	0.711066	0.995696
R2_14	0.719569	1	0.983239	1	0.944448	0.719569	0.991507
R2_150	0.833526	1	0.998103	1	0.0494174	0.833526	0.997699
R2_151	0.550744	1	0.991131	1	0.000491875	0.550744	0.998414
R2_152	0.603368	1	0.997909	1	0.00507309	0.603368	0.998384
R2_153	0.683374	1	0.997373	1	0.980994	0.683374	0.999553
R2_154	0.855658	1	0.993276	1	0.208654	0.855658	0.994631
R2_18	0.628326	1	0.982681	1	0.921885	0.628326	0.986486
R2_19	0.811146	1	0.993283	1	0.517208	0.811146	0.998574
R2_20	0.73318	1	0.999189	1	0.878936	0.73318	0.997907
R2_24	0.586003	0.979592	0.992228	0.957447	1	0.586003	0.998239
R2_26	0.916294	1	0.993199	1	0.943673	0.916294	0.997567
R2_27	0.746429	1	0.995575	1	1	0.746429	0.998805
R2_28	0.688675	1	1	1	1	0.688675	1
R2_32	0.667795	1	0.994107	1	0.998188	0.667795	0.99862
R2_33	0.551237	1	1	1	0.00267738	0.551237	1
R2_37	0.719692	1	0.996328	1	1	0.719692	0.999371
R2_38	0.395517	1	0.985272	1	1	0.395517	0.997506
R2_41	0.838456	1	0.993427	1	0.657222	0.838456	0.989873
R2_42	0.866667	1	1	1	1	0.866667	1
R2_45	0.457304	1	0.998857	1	0.338757	0.457304	0.998423
R2_46	0.708358	1	1	1	1	0.708358	1
R2_47	0.889107	1	0.996537	1	0.994835	0.889107	0.997578
R2_53	0.387345	1	0.996615	1	0.994531	0.387345	0.999075
R2_54	0.864703	1	0.99145	1	0.990187	0.864703	0.997191
R2_55	0.880804	1	0.981338	1	0.981419	0.880804	0.990647
R2_57	0.641837	1	0.993761	0.941176	0.0167529	0.641837	0.998199
R2_59	0.59593	1	0.997914	1	1	0.59593	0.999729
R2_6	0.702025	1	0.996081	1	1	0.702025	0.990476
R2_60	0.608603	1	0.997823	1	0.372401	0.608603	0.995928
R2_63	0.886991	1	0.998307	1	0.998181	0.886991	0.995575
R2_65	1	1	--	--	0.995595	1	0.995595
R2_66	0.527107	1	0.999681	1	0.998892	0.527107	0.992922
R2_67	0.967495	1	1	1	1	0.967495	1
R2_68	0.598525	0.923077	0.996258	0.875	0.998003	0.598525	0.95508
R2_7	0.76132	1	0.998201	1	0.999497	0.76132	0.994472
R2_71	0.780749	1	0.994604	1	1	0.780749	0.999145
R2_72	0.318634	0.866667	0.991273	0.454545	1	0.318634	0.998604
R2_73	0.949033	1	0.996016	1	0.997414	0.949033	0.998803
R2_76	0.473913	1	0.991788	1	1	0.473913	0.999236

```


\pagebreak


\pagebreak


# TransQuin Expression

## Expression summary statistics for: A1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    4.39067 gene

   ***
   *** Detection Limits
   ***

   Break: 3.77655 (R1_62)

   Left:  4.39067 + 0.0805742x (R2 = 0.0567335)
   Right: 2.0794 + 1.00352x (R2 = 0.922502)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.962825
   Slope:       6.46625
   R2:          0.927032
   F-statistic: 902.032
   P-value:     0
   SSM:         3.50824e+10, DF: 1
   SSE:         2.76138e+09, DF: 71
   SST:         3.78438e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.936306
   Slope:       0.724913
   R2:          0.876669
   F-statistic: 504.686
   P-value:     0
   SSM:         1040.46, DF: 1
   SSE:         146.373, DF: 71
   SST:         1186.83, DF: 72


```


\pagebreak


## Expression summary statistics for: A2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    4.62321 gene

   ***
   *** Detection Limits
   ***

   Break: 3.77655 (R1_62)

   Left:  4.62321 + 0.0227324x (R2 = 0.00360572)
   Right: 2.20463 + 0.990951x (R2 = 0.91596)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.960008
   Slope:       6.60733
   R2:          0.921615
   F-statistic: 834.781
   P-value:     0
   SSM:         3.663e+10, DF: 1
   SSE:         3.11546e+09, DF: 71
   SST:         3.97454e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.925689
   Slope:       0.699133
   R2:          0.8569
   F-statistic: 425.155
   P-value:     0
   SSM:         967.772, DF: 1
   SSE:         161.616, DF: 71
   SST:         1129.39, DF: 72


```


\pagebreak


## Expression summary statistics for: A3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking

   Experiment:  60569 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    5.25776 gene

   ***
   *** Detection Limits
   ***

   Break: 15.1062 (R1_101)

   Left:  5.25776 + 0.16804x (R2 = 0.139073)
   Right: 1.51144 + 1.05768x (R2 = 0.903529)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.961001
   Slope:       6.31637
   R2:          0.923522
   F-statistic: 857.371
   P-value:     0
   SSM:         3.34749e+10, DF: 1
   SSE:         2.7721e+09, DF: 71
   SST:         3.6247e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.915017
   Slope:       0.666601
   R2:          0.837255
   F-statistic: 365.266
   P-value:     0
   SSM:         879.802, DF: 1
   SSE:         171.015, DF: 71
   SST:         1050.82, DF: 72


```


\pagebreak


## Expression summary statistics for: B1

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    4.39067 gene

   ***
   *** Detection Limits
   ***

   Break: 3.77655 (R1_62)

   Left:  4.39067 + 0.0805742x (R2 = 0.0567335)
   Right: 2.0794 + 1.00352x (R2 = 0.922502)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.962825
   Slope:       6.46625
   R2:          0.927032
   F-statistic: 902.032
   P-value:     0
   SSM:         3.50824e+10, DF: 1
   SSE:         2.76138e+09, DF: 71
   SST:         3.78438e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.936306
   Slope:       0.724913
   R2:          0.876669
   F-statistic: 504.686
   P-value:     0
   SSM:         1040.46, DF: 1
   SSE:         146.373, DF: 71
   SST:         1186.83, DF: 72


```


\pagebreak


## Expression summary statistics for: B2

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking

   Experiment:  60570 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    4.62321 gene

   ***
   *** Detection Limits
   ***

   Break: 3.77655 (R1_62)

   Left:  4.62321 + 0.0227324x (R2 = 0.00360572)
   Right: 2.20463 + 0.990951x (R2 = 0.91596)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.960008
   Slope:       6.60733
   R2:          0.921615
   F-statistic: 834.781
   P-value:     0
   SSM:         3.663e+10, DF: 1
   SSE:         3.11546e+09, DF: 71
   SST:         3.97454e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.925689
   Slope:       0.699133
   R2:          0.8569
   F-statistic: 425.155
   P-value:     0
   SSM:         967.772, DF: 1
   SSE:         161.616, DF: 71
   SST:         1129.39, DF: 72


```


\pagebreak


## Expression summary statistics for: B3

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking

   Experiment:  60569 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    5.25776 gene

   ***
   *** Detection Limits
   ***

   Break: 15.1062 (R1_101)

   Left:  5.25776 + 0.16804x (R2 = 0.139073)
   Right: 1.51144 + 1.05768x (R2 = 0.903529)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.961001
   Slope:       6.31637
   R2:          0.923522
   F-statistic: 857.371
   P-value:     0
   SSM:         3.34749e+10, DF: 1
   SSE:         2.7721e+09, DF: 71
   SST:         3.6247e+10, DF: 72

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.915017
   Slope:       0.666601
   R2:          0.837255
   F-statistic: 365.266
   P-value:     0
   SSM:         879.802, DF: 1
   SSE:         171.015, DF: 71
   SST:         1050.82, DF: 72


```


\pagebreak


## Expression scatter plot for: A1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(86.9486, 96.745, 3758.17, 1047.79, 55.1694, 37407.9, 3300.12, 622.477, 93.213, 1931.88, 1347.26, 150.979, 15.2912, 32651.3, 72767, 1778.63, 2401.03, 16.2944, 46.4194, 37.6703, 12.6366, 28808.5, 117754, 20.1188, 7491.47, 362.366, 20204.1, 167.12, 18.2308, 667.599, 185.609, 11.608, 8.38523, 1019.99, 49.3218, 347.963, 117035, 5955.88, 15.4836, 149.082, 29.0186, 21487.6, 72072.4, 12446.1, 104.5, 44.0285, 8276.94, 48.9987, 22.042, 26.6032, 6.03984, 23.9704, 15.9987, 1530.13, 36.0393, 39.259, 20.7601, 588.915, 25.2341, 6244.65, 59.6368, 14.562, 255.073, 69.9969, 6775.15, 17.3833, 21.8608, 18.929, 3667.45, 40.9916, 18.488, 20.5209, 20.7893))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression scatter plot for: A2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(85.435, 100.905, 3736.98, 1149.2, 59.4362, 38394.7, 3263.84, 683.137, 108.678, 1937.09, 1344.47, 158.068, 20.6353, 32295.2, 74628.3, 1758.79, 2372.79, 22.4385, 40.1739, 45.5986, 12.1864, 28166.8, 121758, 23.2676, 6764.15, 385.518, 19101.5, 180.522, 24.0939, 611.081, 188.008, 9.74278, 9.02527, 988.693, 49.3037, 342.131, 120242, 5175.77, 18.1181, 158.752, 33.4466, 21403.7, 72627.1, 12062.2, 109.784, 42.817, 8092.31, 55.5725, 32.384, 39.6126, 5.6702, 28.9628, 24.6661, 1510.76, 37.5371, 52.0678, 27.9532, 581.666, 35.8495, 6304.42, 66.5508, 20.6696, 257.151, 85.2934, 6899.09, 18.2791, 20.4355, 21.3052, 3849.71, 47.3217, 23.2432, 25.9138, 33.175))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression scatter plot for: A3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(84.6473, 111.791, 3773.37, 1055.87, 67.8717, 37398.6, 3123.1, 764.926, 111.389, 1910.16, 1331.34, 151.185, 32.7605, 31802.4, 72780.9, 1796.11, 2620.05, 27.1731, 48.5226, 51.0422, 22.3246, 28284.1, 116089, 31.4486, 7029.58, 363.326, 19387.3, 180.159, 31.3369, 622.701, 192.468, 19.0192, 7.2308, 999.325, 83.9476, 353.492, 113321, 5403.58, 21.2101, 163.727, 48.6509, 21459.4, 69322.3, 12056.9, 117.322, 52.7741, 8141.63, 61.1288, 48.8798, 48.6792, 2.43974, 46.0854, 45.5863, 1508.92, 56.355, 83.8905, 43.7215, 579.118, 54.5987, 6389.2, 74.9619, 27.9453, 315.05, 104.529, 6697.33, 17.0121, 30.9176, 23.5949, 3704.14, 61.0756, 38.8357, 36.411, 38.0141))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression scatter plot for: B1

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(86.9486, 96.745, 3758.17, 1047.79, 55.1694, 37407.9, 3300.12, 622.477, 93.213, 1931.88, 1347.26, 150.979, 15.2912, 32651.3, 72767, 1778.63, 2401.03, 16.2944, 46.4194, 37.6703, 12.6366, 28808.5, 117754, 20.1188, 7491.47, 362.366, 20204.1, 167.12, 18.2308, 667.599, 185.609, 11.608, 8.38523, 1019.99, 49.3218, 347.963, 117035, 5955.88, 15.4836, 149.082, 29.0186, 21487.6, 72072.4, 12446.1, 104.5, 44.0285, 8276.94, 48.9987, 22.042, 26.6032, 6.03984, 23.9704, 15.9987, 1530.13, 36.0393, 39.259, 20.7601, 588.915, 25.2341, 6244.65, 59.6368, 14.562, 255.073, 69.9969, 6775.15, 17.3833, 21.8608, 18.929, 3667.45, 40.9916, 18.488, 20.5209, 20.7893))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression scatter plot for: B2

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(85.435, 100.905, 3736.98, 1149.2, 59.4362, 38394.7, 3263.84, 683.137, 108.678, 1937.09, 1344.47, 158.068, 20.6353, 32295.2, 74628.3, 1758.79, 2372.79, 22.4385, 40.1739, 45.5986, 12.1864, 28166.8, 121758, 23.2676, 6764.15, 385.518, 19101.5, 180.522, 24.0939, 611.081, 188.008, 9.74278, 9.02527, 988.693, 49.3037, 342.131, 120242, 5175.77, 18.1181, 158.752, 33.4466, 21403.7, 72627.1, 12062.2, 109.784, 42.817, 8092.31, 55.5725, 32.384, 39.6126, 5.6702, 28.9628, 24.6661, 1510.76, 37.5371, 52.0678, 27.9532, 581.666, 35.8495, 6304.42, 66.5508, 20.6696, 257.151, 85.2934, 6899.09, 18.2791, 20.4355, 21.3052, 3849.71, 47.3217, 23.2432, 25.9138, 33.175))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


## Expression scatter plot for: B3

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 09:55:53.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransExpress -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft Cufflinks -ufiles /Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A1/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A2/genes.fpkm_tracking,/Users/tedwong/Desktop/K_562/Cufflinks/A3/genes.fpkm_tracking 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs     = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_33','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  expected = c(15.1062, 15.1062, 966.797, 241.699, 30.2124, 7734.38, 483.398, 483.398, 15.1062, 483.398, 241.699, 60.4248, 0.118017, 7734.37, 7734.38, 120.85, 1933.59, 0.944138, 120.85, 7.5531, 3.77655, 3867.19, 15468.8, 1.88828, 1933.59, 120.85, 3867.19, 30.2124, 0.472069, 241.699, 60.4248, 0.944138, 0.944138, 120.85, 1.88828, 60.4248, 15468.8, 1933.59, 1.88828, 30.2124, 0.944138, 3867.19, 15468.8, 3867.19, 30.2124, 15.1062, 1933.59, 7.5531, 0.472069, 0.472069, 0.0590086, 0.236034, 0.0590086, 241.699, 0.236034, 0.472069, 0.118017, 120.85, 0.118017, 483.398, 7.5531, 0.472069, 483.398, 7.5531, 966.797, 3.77655, 3.77655, 3.77655, 966.797, 3.77655, 0.236034, 1.88828, 0.0590086),
                  measured = c(84.6473, 111.791, 3773.37, 1055.87, 67.8717, 37398.6, 3123.1, 764.926, 111.389, 1910.16, 1331.34, 151.185, 32.7605, 31802.4, 72780.9, 1796.11, 2620.05, 27.1731, 48.5226, 51.0422, 22.3246, 28284.1, 116089, 31.4486, 7029.58, 363.326, 19387.3, 180.159, 31.3369, 622.701, 192.468, 19.0192, 7.2308, 999.325, 83.9476, 353.492, 113321, 5403.58, 21.2101, 163.727, 48.6509, 21459.4, 69322.3, 12056.9, 117.322, 52.7741, 8141.63, 61.1288, 48.8798, 48.6792, 2.43974, 46.0854, 45.5863, 1508.92, 56.355, 83.8905, 43.7215, 579.118, 54.5987, 6389.2, 74.9619, 27.9453, 315.05, 104.529, 6697.33, 17.0121, 30.9176, 23.5949, 3704.14, 61.0756, 38.8357, 36.411, 38.0141))

plotScatter(data, xname = 'Expected concentration (log2 attomol/ul)', yname = 'Measured coverage (log2 FPKM)')

```


\pagebreak


\pagebreak


# TransQuin Differential

## Differential summary statistics

```{ eval=FALSE}
Summary for dataset: /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv

   Experiment:  60500 gene
   Synthetic:   75 gene

   Reference:   76 gene
   Detected:    75 gene

   ***
   *** Detection Limits
   ***

   Absolute:    0.0590086 (attomol/ul) (R2_38)

   ***
   *** Statistics for linear regression
   ***

   Correlation: 0.707428
   Slope:       0.426606
   R2:          0.500454
   F-statistic: 73.1328
   P-value:     1.29274e-12
   SSM:         302.202, DF: 1
   SSE:         301.653, DF: 73
   SST:         603.855, DF: 74

   ***
   *** Statistics for linear regression (log2 scale)
   ***

   Correlation: 0.790229
   Slope:       0.648599
   R2:          0.624462
   F-statistic: 121.388
   P-value:     0
   SSM:         183.058, DF: 1
   SSE:         110.087, DF: 73
   SST:         293.145, DF: 74


```


\pagebreak


## ROC plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 10:34:52.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransDiff -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft DESeq2 -csoft HTSeqCount -level Gene -cfiles /Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA3v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB3v2.htseq.counts.combined -ufiles /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv 
#

library(Anaquin)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs = c('R1_101','R1_102','R1_103','R1_11','R1_12','R1_13','R1_14','R1_21','R1_22','R1_23','R1_24','R1_31','R1_32','R1_33','R1_41','R1_42','R1_43','R1_51','R1_52','R1_53','R1_61','R1_62','R1_63','R1_71','R1_72','R1_73','R1_81','R1_82','R1_83','R1_91','R1_92','R1_93','R2_1','R2_105','R2_115','R2_116','R2_117','R2_14','R2_150','R2_151','R2_152','R2_153','R2_154','R2_18','R2_19','R2_20','R2_24','R2_26','R2_27','R2_28','R2_32','R2_37','R2_38','R2_41','R2_42','R2_45','R2_46','R2_47','R2_53','R2_54','R2_55','R2_57','R2_59','R2_6','R2_60','R2_63','R2_65','R2_66','R2_67','R2_68','R2_7','R2_71','R2_72','R2_73','R2_76'),
                  pval = c(0.0520723, 0.00114072, 6.25e-27, 7.49e-20, 0.175439, 0.59811, 8.56e-10, 0.266285, 1.85e-21, 0.929809, 1.22e-205, 1.55138e-247, 0.000216566, 0.954847, 4.63e-169, 0, 0, 2.19e-51, 0.000455818, 0.329628, 0.280646, 0.869632, 4.01e-227, 0.923951, 0.601793, 2.48e-210, 0.897165, 0.00420847, 2.51e-15, 0.0706309, 0.00544588, 6.46e-36, 0.0345875, 0.468073, 6.22e-84, 0.576496, 0.425398, 1.54e-49, 4.36e-148, 0.438678, 0.555108, 0.448449, 2.05e-31, 0.134721, 5.66999e-297, 0.846458, 0.347465, 0, 0.0112088, 0.978386, 0.0444392, 0.459634, 0.551343, 1.1e-48, 0.85096, 0.308861, 0.63098, 1.75484e-248, 0.755296, 0.0864418, 5.19e-123, 0.0117735, 0.218838, 1.26e-18, 3.36e-21, 8.73e-40, 0.194597, 7.91e-198, 0.294138, 7.73e-14, 0.407513, 0.358588, 0.319528, 0.333173, 0.11141))

# Classify whether they're TP or FP
data <- transClassify(data)

plotROC(data)

```


\pagebreak


## MA plot

```{r results='hide', message=FALSE, warning=FALSE, echo=FALSE}
#
# Anaquin - Sequin statistical analysis. Version 1.0.
#
# This R script was generated at 19-01-2016 10:34:52.
#
#    /Users/tedwong/Sources/QA/anaquin -t TransDiff -m /Users/tedwong/Desktop/K_562/MTR004.v013.csv -rgtf /Users/tedwong/Desktop/K_562/ATR001.v032.gtf -o /Users/tedwong/Sources/QA/__temp__ -factors 1,1,1,2,2,2 -names A1,A2,A3,B1,B2,B3 -soft DESeq2 -csoft HTSeqCount -level Gene -cfiles /Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/K_RMXA3v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB1v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB2v2.htseq.counts.combined,/Users/tedwong/Desktop/K_562/HTSeqCount/G_RMXB3v2.htseq.counts.combined -ufiles /Users/tedwong/Desktop/K_562/DESeq2/DESeq2.csv 
#

library(Anaquin)

setwd('/Users/tedwong/Sources/QA/__temp__')

# Read the count table
data <- read.csv('TransDiffs_counts.csv', row.name=1)

# Create a TransQuin data set for Anaquin
data <- transQuin(seqs = row.names(data), A1=data$A1, A2=data$A2, A3=data$A3, B1=data$B1, B2=data$B2, B3=data$B3)

plotMA(data, lvl='gene', shouldEndo=TRUE)

```


\pagebreak

