---
title: "Anaquin TransQuin Report"
header-includes: \usepackage{graphicx}
author: "Ted Wong"
date: "8 July 2015"
output: 
    pdf_document:
        keep_tex: true
---

_This is an automatic generation by Anaquin._

# Introduction

An accurate quantification of variability in RNA sequencing analysis is critical for bioinformatics analysis. To control for these sources of variabiltiy, [Garvan Institute of Medical Research](www.garvan.org.au) has developed a set of spiked-in synthetic sequins known as _TransQuin_. The sequins emulate a diverse range of genomic features and they are combined together across a range of concentration to formulate a mixture to emulate quantitative features.    

The sequins align to artificial gene loci encoded within the _in silico_ chromosome. These artificial gene loci are designed to emulate human gene loci including the range of sizes, sequence elements and splicing complexity. The alignment of sequenced reads to these artificial gene loci thereby enables the assessment of RNA sequencing performance and analysis, including spliced-read alignment, isoform assembly, gene discovery and quantitiative gene expression profiling.

_Anaquin_ is a statistical software toolkit built for TransQuin. It is designed to analyze outputs generated by the NGS tools, and accommodates the analysis with detailed statistics on the sequins. Please refer to [www.sequin.xyz](www.sequin.xyz) for more details.

# Data

The datasets included in this study were obtained from two different Illumina sequencing machines, differing in their read length and overall throughput but sharing the same sequencing technology that takes place on a glass slide called a 'flow cell'. A flow cell is made up of eight independent sequencing areas, or 'lanes'. Libraries are deposited on these lanes in order to be sequenced. A library contains cDNAs representative of the RNA molecules that are extracted form a given culture or tissue and are pre-processed in order to be adapted to the sequencing procedire. Simularly to microarrays, the library composition reflects the RNA repertoire expressed in the corresponding culture or tissue. The 'library size' refers to the number of mapped short reads obtained from the sequencing process of the library. In this study, a single library was sequenced in each lane.

# Methods

_K562_ was provided by Gillian Lehrbach (Tissue Culture Facility, The Kinghorn Cancer Centre & Cancer Research Program) and _GM12878_ was provided by Madhavi Maddugoda (Epigenetics Research Group, Garvan Institute of Medical Research). K562 and GM12878 cells were cultured according to Coriell Cell Repositories growth protocols and standards. Briefly, K562 and GM12878 were cultured in RPMI 1640 medium (Gibco®) supplemented with 10% fetal bovine serum (FBS) at 37°C under 5% CO2. 

Total RNA was extracted from K562, GM12878 and mouse using TRIzol (Invitrogen) according to the manufacturer’s instruction. DNAse treatment was subsequently performed on each sample with TURBO DNase (Life Technologies) followed by a cleanup with the RNA Clean and Concentrator-25 Kit (Zymo Research). Total RNA was run on an Agilent Bioanalyzer 2100 to assess intactness and both the Nanodrop (Thermo Scientific) and Qubit (Life Technologies) were used to determine the concentration. Only RNA with a RNA integrity number (RIN) > 8.0 was used for library preparation. 

All 164 RNA transcripts’ concentration was measured on a Qubit 2.0 Fluorometer  (Life Technologies, Carlsbad, CA, USA). The RNA standards mixtures are manufactured in two formulations, RNA Mixture A and RNA Mixture B, each containing the full complement of 164 transcripts. The 164 RNA transcripts in each mixture were pooled using an epMotion 5070 epBlue™ software program to make the final mixtures robotically spanning a 106 -fold concentration range. The transcripts in Mix A and Mix B are present at defined Mix A:Mix B molar concentration ratios, described by 9 subgroups at the gene level and 19 subgroups at the isoform level. 

\newpage

# Reads Alignment

There are 500,456,345 total reads for the six replicates, in which 98% of the reads are mapped. **Figure 1** shows the dilution for each replicate, the average dilution is 2.25%.

\begin{center}
\includegraphics{7.jpeg}
\end{center}

The performance for the alignments at the base, exon and intron level is tablulated in **Figure 2**. An alignment is considered as a true-positive at the exon level if is is aligned within an exon, intron is similar. In the base level, an aligned nucleotide is a true-positive if it is part of an exon, otherwise it is a false-positive.

\begin{center}
\includegraphics{6.png}
\end{center}

**Figure 3** shows the undetected sequins.

\begin{center}
\includegraphics{10.png}
\end{center}

\newpage

# Assembly

The performance for the assembly is compared to the reference TransQuin annotation. This is done at both the isoform and gene level.

\begin{center}
\includegraphics{14.png}
\end{center}

\newpage

# Expression

Gene expression levels and alternative splicing are dynamically regulated during development and between different tissue- or-cell types. By adding alternative RNA sequin mixtures to multiple samples enables a closer comparison between samples, and assessment of differential gene expression and splicing. We repeat the analysis for the synthetic isoforms and genes, as shown in **Figure 3, 4**.

Out of `160` sequins, `150` of them are detected. The undetected sequins are not included in the analysis.

\begin{center}
\includegraphics{9.png}
\end{center}

\begin{center}
\includegraphics{8.png}
\includegraphics{8.png}
\end{center}

**Figure 5** shows the undetected sequins.

\begin{center}
\includegraphics{10.png}
\end{center}

\newpage

# Differential Analysis

The observed TransQuin known-measured relationship is intended for quantitative assessment in differential analysis. A linear regression and it's 95% confidence interval is also plotted.

\begin{center}
\includegraphics{4.png}
\end{center}

\begin{center}
\includegraphics{8.png}
\includegraphics{8.png}
\end{center}

For RNA-Seq experiments, differential expression testing of sequins and endogenous genes was performed with `DESeq2`, to generate P-values for all endogenous and ERCC features. To construct the ROC curves, the 1:1 subpool P-values were the true-negative group for each differential ratio ROC 
curve.

\begin{center}
\textbf{ROC Plot}
\end{center}

\begin{center}
\includegraphics{11.png}
\end{center}

\begin{center}
\textbf{MA Plot}
\end{center}

\begin{center}
\includegraphics{12.png}
\end{center}

\begin{center}
{\Large \textbf{Undetected Sequins}}
\end{center}

\begin{center}
\includegraphics{10.png}
\end{center}

\newpage

# Apprendix

\begin{center}
{\Large \textbf{Density plot for}}
\end{center}

\begin{center}
\includegraphics{1.jpeg}
\end{center}

\begin{center}
{\Large \textbf{Density plot}}
\end{center}

\begin{center}
\includegraphics{1.jpeg}
\end{center}

\begin{center}
{\Large \textbf{Density plot}}
\end{center}

\begin{center}
\includegraphics{1.jpeg}
\end{center}

\begin{center}
{\Large \textbf{Density plot}}
\end{center}

\begin{center}
\includegraphics{1.jpeg}
\end{center}
